#' Write code for server.R
#'
#' @rdname wrSVlib
#' @export wrSVlib
#'
wrSVlib <- function() {
  glue::glue(
    'library(shiny) \n',
    'library(shinyhelper) \n',
    'library(data.table) \n',
    'library(Matrix) \n',
    'library(DT) \n',
    'library(magrittr) \n',
    'library(ggplot2) \n',
    'library(ggrepel) \n',
    'library(hdf5r) \n',
    'library(ggdendro) \n',
    'library(grid) \n',
    'library(gridExtra) \n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVload
#' @export wrSVload
#'
wrSVload <- function(prefix) {
  glue::glue(
    '{prefix}conf = readRDS("./{prefix}conf.rds")\n',
    '{prefix}meta = readRDS("./{prefix}meta.rds")\n',
    '{prefix}gene = readRDS("./{prefix}gene.rds")\n',
    '{prefix}dimr = readRDS("./{prefix}dimr.rds")\n',
    '{prefix}def  = readRDS("./{prefix}def.rds")\n',
    '\n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrSVloadS1
#' @export wrSVloadS1
#'
wrSVloadS1 <- function(prefix) {
  glue::glue(
    '{prefix}image = readRDS("./{prefix}image.rds")\n',
    '\n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrSVloadT1
#' @export wrSVloadT1
#'
wrSVloadT1 <- function(prefix) {
  glue::glue(
    '{prefix}bw   = readRDS("./{prefix}bw.rds")\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVpre
#' @export wrSVpre
#'
wrSVpre <- function() {
  glue::glue(
    'source("./shinyFunc.R")\n',
    'pdf(file = NULL)\n',
    '\n',
    '\n',
    '\n',
    '### Useful stuff \n',
    '# Colour palette \n',
    'cList = list(c("grey85","#FFF7EC","#FEE8C8","#FDD49E","#FDBB84", \n',
    '               "#FC8D59","#EF6548","#D7301F","#B30000","#7F0000"), \n',
    '             c("#4575B4","#74ADD1","#ABD9E9","#E0F3F8","#FFFFBF", \n',
    '               "#FEE090","#FDAE61","#F46D43","#D73027")[c(1,1:9,9)], \n',
    '             c("#FDE725","#AADC32","#5DC863","#27AD81","#21908C", \n',
    '               "#2C728E","#3B528B","#472D7B","#440154")) \n',
    'names(cList) = c("White-Red", "Blue-Yellow-Red", "Yellow-Green-Purple") \n',
    '\n',
    '# Panel sizes \n',
    'pList = c("400px", "600px", "800px") \n',
    'names(pList) = c("Small", "Medium", "Large") \n',
    'pList2 = c("500px", "700px", "900px") \n',
    'names(pList2) = c("Small", "Medium", "Large") \n',
    'pList3 = c("600px", "800px", "1000px") \n',
    'names(pList3) = c("Small", "Medium", "Large") \n',
    'sList = c(18,24,30)\n',
    'names(sList) = c("Small", "Medium", "Large") \n',
    '\n',
    '\n',
    '\n',
    '### Start server code \n',
    'shinyServer(function(input, output, session) {{ \n',
    '  ### For all tags and Server-side selectize \n',
    '  observe_helpers() \n',
    '  optCrt="{{ option_create: function(data,escape) {{return(\'<div class=\\"create\\"><strong>\' + \'</strong></div>\');}} }}" \n', 
    '  observe({{ \n', 
    '    invalidateLater(30000) # ping every 30 seconds to keep connection alive \n', 
    '    cat(".") \n', 
    '  }}) \n', 
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainA1
#' @export wrSVmainA1
#'
wrSVmainA1 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab A1 \n',
    '  getG{prefix}a1inp1 <- reactive({{ \n',
    '    req(gsub("^Assay: ", "", input${prefix}a1ass1)) \n',
    '    if(gsub("^Assay: ", "", input${prefix}a1ass1) == "Cell Information"){{ \n',
    '      res <- {prefix}conf$UI \n',
    '      resDef <- {prefix}def$meta1 \n',
    '      resLen <- length(res) \n',
    '    }} else {{ \n',
    '      res <- names({prefix}gene[[gsub("^Assay: ", "", input${prefix}a1ass1)]]) \n',
    '      resDef <- {prefix}def$gene1[[gsub("^Assay: ", "", input${prefix}a1ass1)]] \n',
    '      resLen <- 7 \n',
    '    }} \n',
    '    return(list(res,resDef,resLen))\n',
    '  }})\n',
    '  observeEvent(gsub("^Assay: ", "", input${prefix}a1ass1), {{\n',
    '    updateSelectizeInput(session, "{prefix}a1inp1", choices = getG{prefix}a1inp1()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}a1inp1()[[2]], options = list( \n',
    '                           maxOptions = getG{prefix}a1inp1()[[3]], create = TRUE, \n',
    '                           persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '  output${prefix}a1sub1.ui <- renderUI({{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a1sub1]$fID, "\\\\|")[[1]] \n',
    '    checkboxGroupInput("{prefix}a1sub2", "Select which cells to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}a1sub1non, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a1sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}a1sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = NULL, inline = TRUE) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}a1sub1all, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a1sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}a1sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = sub, inline = TRUE) \n',
    '  }}) \n',
    '  \n',
    '  {prefix}a1oup1xy <- reactiveValues(x = NULL, y = NULL)\n',
    '  observe({{\n',
    '    brush <- input${prefix}a1inp1.br\n',
    '    if (!is.null(brush)) {{\n',
    '      {prefix}a1oup1xy$x <- c(brush$xmin, brush$xmax)\n',
    '      {prefix}a1oup1xy$y <- c(brush$ymin, brush$ymax)\n',
    '    }} else {{\n',
    '      {prefix}a1oup1xy$x <- NULL; {prefix}a1oup1xy$y <- NULL\n',
    '    }}\n',
    '  }})\n',
    '  {prefix}a1oup1br <- reactive({{\n',
    '    sc2Ddimr({prefix}conf, {prefix}meta, {prefix}dimr, input${prefix}a1dr, input${prefix}a1inp1,\n',
    '             "{prefix}assay_", {prefix}gene, input${prefix}a1ass1, input${prefix}a1sub1, input${prefix}a1sub2,  \n',
    '             input${prefix}a1min1, input${prefix}a1max1, input${prefix}a1siz/2, input${prefix}a1ord1,\n',
    '             cList[[input${prefix}a1col1]], sList[input${prefix}a1fsz]/2, \n',
    '             input${prefix}a1asp, FALSE, FALSE) \n',
    '  }})\n',
    '  output${prefix}a1oup1.br <- renderPlot({{{prefix}a1oup1br() + theme(legend.position = "none")}}) \n',
    '  \n',
    '  {prefix}a1oup1 <- reactive({{\n',
    '    sc2Ddimr({prefix}conf, {prefix}meta, {prefix}dimr, input${prefix}a1dr, input${prefix}a1inp1,\n',
    '             "{prefix}assay_", {prefix}gene, input${prefix}a1ass1, input${prefix}a1sub1, input${prefix}a1sub2,  \n',
    '             input${prefix}a1min1, input${prefix}a1max1, input${prefix}a1siz, input${prefix}a1ord1,\n',
    '             cList[[input${prefix}a1col1]], sList[input${prefix}a1fsz], \n',
    '             input${prefix}a1asp, input${prefix}a1txt, input${prefix}a1lab1)\n',
    '  }})\n',
    '  output${prefix}a1oup1 <- renderPlot({{\n',
    '    if(is.null({prefix}a1oup1xy$x[1])){{\n',
    '      {prefix}a1oup1() + theme(legend.position = "none")\n',
    '    }} else {{\n',
    '      {prefix}a1oup1() + theme(legend.position = "none") + \n',
    '        scale_x_continuous(limits = {prefix}a1oup1xy$x, expand = c(0, 0)) + \n',
    '        scale_y_continuous(limits = {prefix}a1oup1xy$y, expand = c(0, 0)) \n',
    '    }}\n',
    '  }})\n',
    '  output${prefix}a1oup1.ui <- renderUI({{plotOutput("{prefix}a1oup1", height = pList[input${prefix}a1psz])}})\n',
    '  output${prefix}a1oup1.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}a1dr,"_",input${prefix}a1inp1,".",input${prefix}a1oup1.f) }},\n',
    '    content = function(file) {{ \n',
    '      if(is.null({prefix}a1oup1xy$x[1])){{\n',
    '        ggsav(file, height = input${prefix}a1oup1.h, width = input${prefix}a1oup1.w, \n',
    '              plot = {prefix}a1oup1() + theme(legend.position = "none"))\n',
    '      }} else {{\n',
    '        ggsav(file, height = input${prefix}a1oup1.h, width = input${prefix}a1oup1.w, \n',
    '              plot = {prefix}a1oup1() + theme(legend.position = "none") + \n',
    '                scale_x_continuous(limits = {prefix}a1oup1xy$x, expand = c(0, 0)) + \n',
    '                scale_y_continuous(limits = {prefix}a1oup1xy$y, expand = c(0, 0)))\n',
    '      }}\n',
    '  }})\n',
    '  output${prefix}a1oup3 <- renderPlot({{grid.newpage(); grid.draw(g_legend({prefix}a1oup1()))}})\n',
    '  output${prefix}a1oup3.ui <- renderUI({{plotOutput("{prefix}a1oup3", height = 72*convertHeight(\n',
    '    grobHeight(g_legend({prefix}a1oup1())), unitTo="in", valueOnly=TRUE) + 50)}})\n',
    '  output${prefix}a1oup3.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}a1dr,"_",input${prefix}a1inp1,"_leg.",input${prefix}a1oup3.f) }},\n',
    '    content = function(file) {{ \n',
    '      grid.newpage(); grid.draw(g_legend({prefix}a1oup1()))\n',
    '      ggsav(file, height = input${prefix}a1oup3.h, width = input${prefix}a1oup3.w, plot = grid.grab())\n',
    '  }}) \n',
    '  \n',
    '  output${prefix}a1.dt <- renderDataTable({{\n',
    '    ggData = sc2Dnum({prefix}conf, {prefix}meta, {prefix}dimr, input${prefix}a1dr, {prefix}a1oup1xy$x, {prefix}a1oup1xy$y, input${prefix}a1inp1,\n',
    '                     "{prefix}assay_", {prefix}gene, input${prefix}a1ass1, input${prefix}a1sub1, input${prefix}a1sub2, input${prefix}a1splt)\n',
    '    datatable(ggData, rownames = FALSE, extensions = "Buttons",\n',
    '              options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>%\n',
    '      formatRound(columns = c("pctZoom"), digits = 2)\n',
    '  }})\n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainA2
#' @export wrSVmainA2
#'
wrSVmainA2 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab A2 \n',
    '  getG{prefix}a2inp1 <- reactive({{ \n',
    '    req(gsub("^Assay: ", "", input${prefix}a2ass1)) \n',
    '    if(gsub("^Assay: ", "", input${prefix}a2ass1) == "Cell Information"){{ \n',
    '      res <- {prefix}conf$UI \n',
    '      resDef <- {prefix}def$meta1 \n',
    '      resLen <- length(res) \n',
    '    }} else {{ \n',
    '      res <- names({prefix}gene[[gsub("^Assay: ", "", input${prefix}a2ass1)]]) \n',
    '      resDef <- {prefix}def$gene1[[gsub("^Assay: ", "", input${prefix}a2ass1)]] \n',
    '      resLen <- 7 \n',
    '    }} \n',
    '    return(list(res,resDef,resLen))\n',
    '  }})\n',
    '  observeEvent(gsub("^Assay: ", "", input${prefix}a2ass1), {{\n',
    '    updateSelectizeInput(session, "{prefix}a2inp1", choices = getG{prefix}a2inp1()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}a2inp1()[[2]], options = list( \n',
    '                           maxOptions = getG{prefix}a2inp1()[[3]], create = TRUE, \n',
    '                           persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '  getG{prefix}a2inp2 <- reactive({{ \n',
    '    req(gsub("^Assay: ", "", input${prefix}a2ass2)) \n',
    '    if(gsub("^Assay: ", "", input${prefix}a2ass2) == "Cell Information"){{ \n',
    '      res <- {prefix}conf$UI \n',
    '      resDef <- {prefix}def$meta1 \n',
    '      resLen <- length(res) \n',
    '    }} else {{ \n',
    '      res <- names({prefix}gene[[gsub("^Assay: ", "", input${prefix}a2ass2)]]) \n',
    '      resDef <- {prefix}def$gene1[[gsub("^Assay: ", "", input${prefix}a2ass2)]] \n',
    '      resLen <- 7 \n',
    '    }} \n',
    '    return(list(res,resDef,resLen))\n',
    '  }})\n',
    '  observeEvent(gsub("^Assay: ", "", input${prefix}a2ass2), {{\n',
    '    updateSelectizeInput(session, "{prefix}a2inp2", choices = getG{prefix}a2inp2()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}a2inp2()[[2]], options = list( \n',
    '                           maxOptions = getG{prefix}a2inp2()[[3]], create = TRUE, \n',
    '                           persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '    output${prefix}a2sub1.ui <- renderUI({{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a2sub1]$fID, "\\\\|")[[1]] \n',
    '    checkboxGroupInput("{prefix}a2sub2", "Select which cells to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}a2sub1non, {{\n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a2sub1]$fID, "\\\\|")[[1]]\n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}a2sub2", label = "Select which cells to show",\n',
    '                             choices = sub, selected = NULL, inline = TRUE)\n',
    '  }})\n',
    '  observeEvent(input${prefix}a2sub1all, {{\n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a2sub1]$fID, "\\\\|")[[1]]\n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}a2sub2", label = "Select which cells to show",\n',
    '                             choices = sub, selected = sub, inline = TRUE)\n',
    '  }})\n',
    '  \n',
    '  {prefix}a2oup1 <- reactive({{\n',
    '    sc2Ddimr({prefix}conf, {prefix}meta, {prefix}dimr, input${prefix}a2dr, input${prefix}a2inp1,\n',
    '             "{prefix}assay_", {prefix}gene, input${prefix}a2ass1, input${prefix}a2sub1, input${prefix}a2sub2,  \n',
    '             input${prefix}a2min1, input${prefix}a2max1, input${prefix}a2siz, input${prefix}a2ord1,\n',
    '             cList[[input${prefix}a2col1]], sList[input${prefix}a2fsz], \n',
    '             input${prefix}a2asp, input${prefix}a2txt, input${prefix}a2lab1) \n',
    '  }})\n',
    '  output${prefix}a2oup1 <- renderPlot({{{prefix}a2oup1() + theme(legend.position = "none")}})\n',
    '  output${prefix}a2oup1.ui <- renderUI({{plotOutput("{prefix}a2oup1", height = pList[input${prefix}a2psz])}})\n',
    '  output${prefix}a2oup1.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}a2dr,"_",input${prefix}a2inp1,".",input${prefix}a2oup1.f) }},\n',
    '    content = function(file) {{ ggsav(file, height = input${prefix}a2oup1.h, width = input${prefix}a2oup1.w, \n',
    '                                     plot = {prefix}a2oup1() + theme(legend.position = "none"))\n',
    '  }})\n',
    '  output${prefix}a2oup3 <- renderPlot({{grid.newpage(); grid.draw(g_legend({prefix}a2oup1()))}})\n',
    '  output${prefix}a2oup3.ui <- renderUI({{plotOutput("{prefix}a2oup3", height = 72*convertHeight(\n',
    '    grobHeight(g_legend({prefix}a2oup1())), unitTo="in", valueOnly=TRUE) + 50)}})\n',
    '  output${prefix}a2oup3.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}a2dr,"_",input${prefix}a2inp1,"_leg.",input${prefix}a2oup3.f) }},\n',
    '  content = function(file) {{ \n',
    '    grid.newpage(); grid.draw(g_legend({prefix}a2oup1()))\n',
    '    ggsav(file, height = input${prefix}a2oup3.h, width = input${prefix}a2oup3.w, plot = grid.grab())\n',
    '  }})\n',
    '  \n',
    '  {prefix}a2oup2 <- reactive({{\n',
    '    sc2Ddimr({prefix}conf, {prefix}meta, {prefix}dimr, input${prefix}a2dr, input${prefix}a2inp2,\n',
    '             "{prefix}assay_", {prefix}gene, input${prefix}a2ass2, input${prefix}a2sub1, input${prefix}a2sub2,  \n',
    '             input${prefix}a2min2, input${prefix}a2max2, input${prefix}a2siz, input${prefix}a2ord2,\n',
    '             cList[[input${prefix}a2col2]], sList[input${prefix}a2fsz], \n',
    '             input${prefix}a2asp, input${prefix}a2txt, input${prefix}a2lab2) \n',
    '  }})\n',
    '  output${prefix}a2oup2 <- renderPlot({{{prefix}a2oup2() + theme(legend.position = "none")}})\n',
    '  output${prefix}a2oup2.ui <- renderUI({{plotOutput("{prefix}a2oup2", height = pList[input${prefix}a2psz])}})\n',
    '  output${prefix}a2oup2.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}a2dr,"_",input${prefix}a2inp2,".",input${prefix}a2oup2.f) }},\n',
    '    content = function(file) {{ ggsav(file, height = input${prefix}a2oup2.h, width = input${prefix}a2oup2.w, \n',
    '                                     plot = {prefix}a2oup2() + theme(legend.position = "none"))\n',
    '  }})\n',
    '  \n',
    '  output${prefix}a2oup4 <- renderPlot({{grid.newpage(); grid.draw(g_legend({prefix}a2oup2()))}})\n',
    '  output${prefix}a2oup4.ui <- renderUI({{plotOutput("{prefix}a2oup4", height = 72*convertHeight(\n',
    '    grobHeight(g_legend({prefix}a2oup2())), unitTo="in", valueOnly=TRUE) + 50)}})\n',
    '  output${prefix}a2oup4.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}a2dr,"_",input${prefix}a2inp2,"_leg.",input${prefix}a2oup4.f) }},\n',
    '  content = function(file) {{ \n',
    '    grid.newpage(); grid.draw(g_legend({prefix}a2oup2()))\n',
    '    ggsav(file, height = input${prefix}a2oup4.h, width = input${prefix}a2oup4.w, plot = grid.grab())\n',
    '  }})\n',
    '  \n',
    '  {prefix}a2oup5 <- reactive({{\n',
    '    sc2Dcomp({prefix}conf, {prefix}meta, input${prefix}a2inp1, input${prefix}a2inp2,\n',
    '             "{prefix}assay_", {prefix}gene, input${prefix}a2ass1, input${prefix}a2ass2, input${prefix}a2sub1, input${prefix}a2sub2,\n',
    '             input${prefix}a2min1, input${prefix}a2max1, input${prefix}a2min2, input${prefix}a2max2, \n',
    '             input${prefix}a2siz, cList[[input${prefix}a2col2]], sList[input${prefix}a2fsz]) \n',
    '  }})\n',
    '  output${prefix}a2oup5 <- renderPlot({{{prefix}a2oup5() + theme(legend.position = "none")}})\n',
    '  output${prefix}a2oup5.ui <- renderUI({{plotOutput("{prefix}a2oup5", height = pList[input${prefix}a2psz])}})\n',
    '  output${prefix}a2oup5.dl <- downloadHandler( \n',
    '    filename = function() {{ paste0("{prefix}compare_",input${prefix}a2inp1,"_",input${prefix}a2inp2,".",input${prefix}a2oup5.f) }, \n',
    '    content = function(file) {{ ggsav(file, height = input${prefix}a2oup5.h, width = input${prefix}a2oup5.w, \n',
    '                                     plot = {prefix}a2oup5()) \n',
    '  }}) \n',
    '  output${prefix}a2.dt <- renderDataTable({{\n',
    '    ggData = sc2Dcnum({prefix}conf, {prefix}meta, input${prefix}a2inp1, input${prefix}a2inp2,\n',
    '                      "{prefix}assay_", {prefix}gene, input${prefix}a2ass1, input${prefix}a2ass2, input${prefix}a2sub1, input${prefix}a2sub2,\n',
    '                      input${prefix}a2min1, input${prefix}a2max1, input${prefix}a2min2, input${prefix}a2max2, input${prefix}a2cut)\n',
    '    datatable(ggData, rownames = FALSE, extensions = "Buttons",\n',
    '              options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel")))\n',
    '  }})\n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainA3
#' @export wrSVmainA3
#'
wrSVmainA3 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab A3 \n',
    '  getG{prefix}a3inp1 <- reactive({{ \n',
    '    req(input${prefix}a3ass1) \n',
    '    res <- names({prefix}gene[[input${prefix}a3ass1]]) \n',
    '    resDef1 <- {prefix}def$gene1[[input${prefix}a3ass1]] \n',
    '    resDef2 <- {prefix}def$gene2[[input${prefix}a3ass1]] \n',
    '    return(list(res,resDef1,resDef2))\n',
    '  }})\n',
    '  observeEvent(input${prefix}a3ass1, {{\n',
    '    updateSelectizeInput(session, "{prefix}a3inp1", choices = getG{prefix}a3inp1()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}a3inp1()[[2]], options = list( \n',
    '                           maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '  observeEvent(input${prefix}a3ass1, {{\n',
    '    updateSelectizeInput(session, "{prefix}a3inp2", choices = getG{prefix}a3inp1()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}a3inp1()[[3]], options = list( \n',
    '                           maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '  output${prefix}a3sub1.ui <- renderUI({{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a3sub1]$fID, "\\\\|")[[1]] \n',
    '    checkboxGroupInput("{prefix}a3sub2", "Select which cells to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}a3sub1non, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a3sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}a3sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = NULL, inline = TRUE) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}a3sub1all, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}a3sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}a3sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = sub, inline = TRUE) \n',
    '  }}) \n',
    '  {prefix}a3oup1 <- reactive({{\n',
    '    scDRcoex({prefix}conf, {prefix}meta, {prefix}dimr, input${prefix}a3dr, input${prefix}a3inp1, input${prefix}a3inp2, \n',
    '             paste0("{prefix}assay_", input${prefix}a3ass1, ".h5"), {prefix}gene[[input${prefix}a3ass1]], \n',
    '             input${prefix}a3sub1, input${prefix}a3sub2, input${prefix}a3min1, input${prefix}a3max1, input${prefix}a3min2, input${prefix}a3max2, \n',
    '             input${prefix}a3siz, input${prefix}a3col1, input${prefix}a3ord1, \n',
    '             sList[input${prefix}a3fsz], input${prefix}a3asp, input${prefix}a3txt) \n',
    '  }}) \n',
    '  output${prefix}a3oup1 <- renderPlot({{{prefix}a3oup1()}}) \n',
    '  output${prefix}a3oup1.ui <- renderUI({{plotOutput("{prefix}a3oup1", height = pList2[input${prefix}a3psz])}}) \n',
    '  output${prefix}a3oup1.dl <- downloadHandler( \n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}a3dr,"_",input${prefix}a3inp1,"_",input${prefix}a3inp2,".",input${prefix}a3oup1.f) }}, \n',
    '    content = function(file) {{ ggsav(file, height = input${prefix}a3oup1.h, width = input${prefix}a3oup1.w, plot = {prefix}a3oup1())\n',
    '  }}) \n',
    '  \n',
    '  output${prefix}a3oup2 <- renderPlot({{scDRcoexLeg(input${prefix}a3inp1, input${prefix}a3inp2, input${prefix}a3col1, sList[input${prefix}a3fsz])}}) \n',
    '  output${prefix}a3oup2.ui <- renderUI({{plotOutput("{prefix}a3oup2", height = "300px")}})\n',
    '  output${prefix}a3oup2.dl <- downloadHandler( \n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}a3dr,"_",input${prefix}a3inp1,"_",input${prefix}a3inp2,"_leg.",input${prefix}a3oup2.f) }}, \n',
    '    content = function(file) {{ ggsav(file, height = 3, width = 4, \n',
    '                                     plot = scDRcoexLeg(input${prefix}a3inp1, input${prefix}a3inp2, input${prefix}a3col1, sList[input${prefix}a3fsz]))\n',
    '  }}) \n',
    '  output${prefix}a3.dt <- renderDataTable({{ \n',
    '    ggData = scDRcoexNum({prefix}conf, {prefix}meta, input${prefix}a3inp1, input${prefix}a3inp2, \n',
    '                         paste0("{prefix}assay_", input${prefix}a3ass1, ".h5"), {prefix}gene[[input${prefix}a3ass1]], \n',
    '                         input${prefix}a3sub1, input${prefix}a3sub2) \n',
    '    datatable(ggData, rownames = FALSE, extensions = "Buttons", \n',
    '              options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>% \n',
    '      formatRound(columns = c("percent"), digits = 2) \n',
    '  }}) \n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainB1
#' @export wrSVmainB1
#'
wrSVmainB1 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab B1 \n',
    '  getG{prefix}b1inp1 <- reactive({{ \n',
    '    req(gsub("^Assay: ", "", input${prefix}b1ass1)) \n',
    '    if(gsub("^Assay: ", "", input${prefix}b1ass1) == "Cell Information"){{ \n',
    '      res <- {prefix}conf[is.na(fID)]$UI \n',
    '      resDef <- {prefix}conf[is.na(fID)]$UI[1] \n',
    '      resLen <- length(res) \n',
    '    }} else {{ \n',
    '      res <- names({prefix}gene[[gsub("^Assay: ", "", input${prefix}b1ass1)]]) \n',
    '      resDef <- {prefix}def$gene1[[gsub("^Assay: ", "", input${prefix}b1ass1)]] \n',
    '      resLen <- 7 \n',
    '    }} \n',
    '    return(list(res,resDef,resLen))\n',
    '  }})\n',
    '  observeEvent(gsub("^Assay: ", "", input${prefix}b1ass1), {{\n',
    '    updateSelectizeInput(session, "{prefix}b1inp2", choices = getG{prefix}b1inp1()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}b1inp1()[[2]], options = list( \n',
    '                           maxOptions = getG{prefix}b1inp1()[[3]], create = TRUE, \n',
    '                           persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '  output${prefix}b1sub1.ui <- renderUI({{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b1sub1]$fID, "\\\\|")[[1]] \n',
    '    checkboxGroupInput("{prefix}b1sub2", "Select which cells to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}b1sub1non, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b1sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}b1sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = NULL, inline = TRUE) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}b1sub1all, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b1sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}b1sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = sub, inline = TRUE) \n',
    '  }}) \n',
    '\n',
    '  {prefix}b1oup <- reactive({{\n',
    '    scVioBox({prefix}conf, {prefix}meta, input${prefix}b1inp1, input${prefix}b1inp2, \n',
    '             "{prefix}assay_", {prefix}gene, input${prefix}b1ass1, \n',
    '             input${prefix}b1sub1, input${prefix}b1sub2, input${prefix}b1typ, input${prefix}b1pts, \n',
    '             input${prefix}b1stg, input${prefix}b1stp1, input${prefix}b1stp2, \n',
    '             input${prefix}b1siz, sList[input${prefix}b1fsz], input${prefix}b1noi) \n',
    '  }}) \n',
    '  output${prefix}b1oup <- renderPlot({{{prefix}b1oup()}}) \n',
    '  output${prefix}b1oup.ui <- renderUI({{plotOutput("{prefix}b1oup", height = pList2[input${prefix}b1psz])}}) \n',
    '  output${prefix}b1oup.dl <- downloadHandler( \n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}b1typ,"_",input${prefix}b1inp1,"_",input${prefix}b1inp2,".",input${prefix}b1oup.f) }}, \n',
    '    content = function(file) {{ ggsave(file, height = input${prefix}b1oup.h, width = input${prefix}b1oup.w, plot = {prefix}b1oup())\n',
    '  }}) \n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainB2
#' @export wrSVmainB2
#'
wrSVmainB2 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab B2 \n',
    '  observeEvent(input${prefix}b2inp1, {{\n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b2inp2]$fID, "\\\\|")[[1]] \n',
    '    updateSelectizeInput(session, "{prefix}b2ord1", choices = c("Original order", sub), \n',
    '                         server = TRUE, selected = "Original order", options = list( \n',
    '                         create = TRUE, persist = TRUE, render = I(optCrt))) \n',
    '  }}) \n',
    '  output${prefix}b2sub1.ui <- renderUI({{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b2sub1]$fID, "\\\\|")[[1]] \n',
    '    checkboxGroupInput("{prefix}b2sub2", "Select which cells to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}b2sub1non, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b2sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}b2sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = NULL, inline = TRUE) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}b2sub1all, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b2sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}b2sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = sub, inline = TRUE) \n',
    '  }}) \n',
    '  \n',
    '  {prefix}b2oup  <- reactive({{\n',
    '    scProp({prefix}conf, {prefix}meta, input${prefix}b2inp1, input${prefix}b2inp2,  \n',
    '           input${prefix}b2sub1, input${prefix}b2sub2, input${prefix}b2ord1, input${prefix}b2ord2, \n',
    '           input${prefix}b2typ, input${prefix}b2flp, sList[input${prefix}b2fsz]) \n',
    '  }})\n',
    '  output${prefix}b2oup <- renderPlot({{{prefix}b2oup()}}) \n',
    '  output${prefix}b2oup.ui <- renderUI({{plotOutput("{prefix}b2oup", height = pList2[input${prefix}b2psz])}}) \n',
    '  output${prefix}b2oup.dl <- downloadHandler( \n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}b2typ,"_",input${prefix}b2inp1,"_",input${prefix}b2inp2,".",input${prefix}b2oup.f) }}, \n',
    '    content = function(file) {{ ggsave(file, height = input${prefix}b2oup.h, width = input${prefix}b2oup.w, plot = {prefix}b2oup())\n',
    '  }}) \n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainB3
#' @export wrSVmainB3
#'
wrSVmainB3 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab B3 \n',
    '  getG{prefix}b3inp1 <- reactive({{ \n',
    '    req(input${prefix}b3ass1) \n',
    '    resDef <- paste0({prefix}def$genes[[input${prefix}b3ass1]], \n',
    '                     collapse = ", ") \n',
    '    return(resDef) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}b3ass1, {{ \n',
    '    updateTextAreaInput(session, "{prefix}b3inp", value = getG{prefix}b3inp1()) \n',
    '  }}) \n',
    '  output${prefix}b3sub1.ui <- renderUI({{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b3sub1]$fID, "\\\\|")[[1]] \n',
    '    checkboxGroupInput("{prefix}b3sub2", "Select which cells to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}b3sub1non, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b3sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}b3sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = NULL, inline = TRUE) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}b3sub1all, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}b3sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}b3sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = sub, inline = TRUE) \n',
    '  }}) \n',
    '  output${prefix}b3oupTxt <- renderUI({{ \n',
    '    geneList = scGeneList(input${prefix}b3inp, {prefix}gene[[input${prefix}b3ass1]]) \n',
    '    if(nrow(geneList) > 50){{ \n',
    '      HTML("More than 50 input genes! Please reduce the gene list!") \n',
    '    }} else {{ \n',
    '      oup = paste0(nrow(geneList[present == TRUE]), " genes OK and will be plotted") \n',
    '      if(nrow(geneList[present == FALSE]) > 0){{ \n',
    '        oup = paste0(oup, "<br/>", \n',
    '                     nrow(geneList[present == FALSE]), " genes not found (", \n',
    '                     paste0(geneList[present == FALSE]$gene, collapse = ", "), ")") \n',
    '      }} \n',
    '      HTML(oup) \n',
    '    }} \n',
    '  }}) \n',
    '  \n',
    '  {prefix}b3oup <- reactive({{\n',
    '    scBubbHeat({prefix}conf, {prefix}meta, input${prefix}b3inp, input${prefix}b3grp, input${prefix}b3plt, \n',
    '               paste0("{prefix}assay_", input${prefix}b3ass1, ".h5"), {prefix}gene[[input${prefix}b3ass1]], \n',
    '               input${prefix}b3sub1, input${prefix}b3sub2, input${prefix}b3scl, \n',
    '               input${prefix}b3row, input${prefix}b3col, cList[[input${prefix}b3cols]], sList[input${prefix}b3fsz], \n',
    '               input${prefix}b3exp, input${prefix}b3max) \n',
    '  }}) \n',
    '  output${prefix}b3oup <- renderPlot({{{prefix}b3oup()}}) \n',
    '  output${prefix}b3oup.ui <- renderUI({{plotOutput("{prefix}b3oup", height = pList3[input${prefix}b3psz])}})\n',
    '  output${prefix}b3oup.dl <- downloadHandler( \n',
    '    filename = function() {{ paste0("{prefix}",input${prefix}b3plt,"_",input${prefix}b3grp,".",input${prefix}b3oup.f) }}, \n',
    '    content = function(file) {{ ggsave(file, height = input${prefix}b3oup.h, width = input${prefix}b3oup.w, plot = {prefix}b3oup())\n',
    '  }}) \n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainS1
#' @export wrSVmainS1
#'
wrSVmainS1 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab S1 \n',
    '  getG{prefix}s1inp1 <- reactive({{ \n',
    '    req(gsub("^Assay: ", "", input${prefix}s1ass1)) \n',
    '    if(gsub("^Assay: ", "", input${prefix}s1ass1) == "Cell Information"){{ \n',
    '      res <- {prefix}conf$UI \n',
    '      resDef <- {prefix}def$meta1 \n',
    '      resLen <- length(res) \n',
    '    }} else {{ \n',
    '      res <- names({prefix}gene[[gsub("^Assay: ", "", input${prefix}s1ass1)]]) \n',
    '      resDef <- {prefix}def$gene1[[gsub("^Assay: ", "", input${prefix}s1ass1)]] \n',
    '      resLen <- 7 \n',
    '    }} \n',
    '    return(list(res,resDef,resLen))\n',
    '  }})\n',
    '  observeEvent(gsub("^Assay: ", "", input${prefix}s1ass1), {{\n',
    '    updateSelectizeInput(session, "{prefix}s1inp1", choices = getG{prefix}s1inp1()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}s1inp1()[[2]], options = list( \n',
    '                           maxOptions = getG{prefix}s1inp1()[[3]], create = TRUE, \n',
    '                           persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '  output${prefix}s1sub1.ui <- renderUI({{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}s1sub1]$fID, "\\\\|")[[1]] \n',
    '    checkboxGroupInput("{prefix}s1sub2", "Select which cells to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}s1sub1non, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}s1sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}s1sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = NULL, inline = TRUE) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}s1sub1all, {{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}s1sub1]$fID, "\\\\|")[[1]] \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}s1sub2", label = "Select which cells to show", \n',
    '                             choices = sub, selected = sub, inline = TRUE) \n',
    '  }}) \n',  
    '  \n',
    '  {prefix}s1oup1xy <- reactiveValues(x = NULL, y = NULL)\n',
    '  observe({{\n',
    '    brush <- input${prefix}s1inp1.br\n',
    '    if (!is.null(brush)) {{\n',
    '      {prefix}s1oup1xy$x <- c(brush$xmin, brush$xmax)\n',
    '      {prefix}s1oup1xy$y <- c(brush$ymin, brush$ymax)\n',
    '    }} else {{\n',
    '      {prefix}s1oup1xy$x <- NULL; {prefix}s1oup1xy$y <- NULL\n',
    '    }}\n',
    '  }})  \n',
    '  {prefix}s1oup1br <- reactive({{\n',
    '    sc2Dspat({prefix}conf, {prefix}meta, {prefix}image, input${prefix}s1alp, \n',
    '             input${prefix}s1inp1, "{prefix}assay_", {prefix}gene, input${prefix}s1ass1, \n',
    '             input${prefix}s1sub1, input${prefix}s1sub2, input${prefix}s1min1, input${prefix}s1max1, input${prefix}s1siz*0.6, \n',
    '             cList[[input${prefix}s1col1]], sList[input${prefix}s1fsz]*0.6, FALSE) \n',
    '  }})\n',
    '  output${prefix}s1oup1.br <- renderPlot({{{prefix}s1oup1br() + theme(legend.position = "none")}}) \n',  
    '  \n',
    '  {prefix}s1oup1 <- reactive({{\n',
    '    sc2Dspat({prefix}conf, {prefix}meta, {prefix}image, input${prefix}s1alp, \n',
    '             input${prefix}s1inp1, "{prefix}assay_", {prefix}gene, input${prefix}s1ass1, \n',
    '             input${prefix}s1sub1, input${prefix}s1sub2, input${prefix}s1min1, input${prefix}s1max1, input${prefix}s1siz, \n',
    '             cList[[input${prefix}s1col1]], sList[input${prefix}s1fsz], input${prefix}s1lab1) \n',
    '  }})\n',
    '  output${prefix}s1oup1 <- renderPlot({{\n',
    '    if(is.null({prefix}s1oup1xy$x[1])){{\n',
    '      {prefix}s1oup1() + theme(legend.position = "none")\n',
    '    }} else {{\n',
    '      rngX = diff(range({prefix}image$coord$imagecol)) / diff({prefix}s1oup1xy$x)\n',
    '      rngY = diff(range({prefix}image$coord$imagerow)) / diff({prefix}s1oup1xy$y)\n',
    '      adjSiz = input${prefix}s1siz * min(rngX, rngY)\n',
    '      {prefix}s1oup1() + theme(legend.position = "none") + \n',
    '        scale_x_continuous(limits = {prefix}s1oup1xy$x, expand = c(0, 0)) + \n',
    '        scale_y_continuous(limits = {prefix}s1oup1xy$y, expand = c(0, 0)) + \n',
    '        scale_size_continuous(range = c(0, adjSiz), limits = c(0,1))\n',
    '    }}\n',
    '  }})\n',
    '  output${prefix}s1oup1.ui <- renderUI({{plotOutput("{prefix}s1oup1", height = pList[input${prefix}s1psz])}})\n',
    '  output${prefix}s1oup1.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}spatial_",input${prefix}s1inp1,".",input${prefix}s1oup1.f) }},\n',
    '    content = function(file) {{ \n',
    '      if(is.null({prefix}s1oup1xy$x[1])){{\n',
    '        ggsav(file, height = input${prefix}s1oup1.h, width = input${prefix}s1oup1.w, \n',
    '              plot = {prefix}s1oup1() + theme(legend.position = "none"))\n',
    '      }} else {{\n',
    '        rngX = diff(range({prefix}image$coord$imagecol)) / diff({prefix}s1oup1xy$x)\n',
    '        rngY = diff(range({prefix}image$coord$imagerow)) / diff({prefix}s1oup1xy$y)\n',
    '        adjSiz = input${prefix}s1siz * min(rngX, rngY)\n',
    '        ggsav(file, height = input${prefix}s1oup1.h, width = input${prefix}s1oup1.w, \n',
    '              plot = {prefix}s1oup1() + theme(legend.position = "none") + \n',
    '                scale_x_continuous(limits = {prefix}s1oup1xy$x, expand = c(0, 0)) + \n',
    '                scale_y_continuous(limits = {prefix}s1oup1xy$y, expand = c(0, 0)) + \n',
    '                scale_size_continuous(range = c(0, adjSiz), limits = c(0,1)))\n',
    '      }}\n',
    '  }})\n',
    '  output${prefix}s1oup3 <- renderPlot({{grid.newpage(); grid.draw(g_legend({prefix}s1oup1()))}})\n',
    '  output${prefix}s1oup3.ui <- renderUI({{plotOutput("{prefix}s1oup3", height = 72*convertHeight(\n',
    '    grobHeight(g_legend({prefix}s1oup1())), unitTo="in", valueOnly=TRUE) + 50)}})\n',
    '  output${prefix}s1oup3.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}spatial_",input${prefix}s1inp1,"_leg.",input${prefix}s1oup3.f) }},\n',
    '    content = function(file) {{ \n',
    '      grid.newpage(); grid.draw(g_legend({prefix}s1oup1()))\n',
    '      ggsav(file, height = input${prefix}s1oup3.h, width = input${prefix}s1oup3.w, plot = grid.grab())\n',
    '    }}) \n',
    '  \n',
    '  output${prefix}s1.dt <- renderDataTable({{\n',
    '    orgXY = c(0, dim({prefix}image[["bg_image"]])[1], 0, dim({prefix}image[["bg_image"]])[2]) \n',
    '    orgXY = orgXY / {prefix}image$lowres \n',
    '    ggData = data.table({prefix}image$coord[, c("imagecol", "imagerow")])\n',
    '    colnames(ggData) = c("UMAP_1","UMAP_2")\n',
    '    ggData$UMAP_2 = orgXY[4] - ggData$UMAP_2\n',
    '    tmpdimr = list(); tmpdimr$umap = ggData\n',
    '    \n',
    '    ggData = sc2Dnum({prefix}conf, {prefix}meta, tmpdimr, "umap", {prefix}s1oup1xy$x, {prefix}s1oup1xy$y, input${prefix}s1inp1,\n',
    '                     "{prefix}assay_", {prefix}gene, input${prefix}s1ass1, input${prefix}s1sub1, input${prefix}s1sub2, input${prefix}s1splt)\n',
    '    datatable(ggData, rownames = FALSE, extensions = "Buttons",\n',
    '              options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>%\n',
    '      formatRound(columns = c("pctZoom"), digits = 2)\n',
    '  }})\n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainS2
#' @export wrSVmainS2
#'
wrSVmainS2 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab S2 \n',
    '  getG{prefix}s2inp1 <- reactive({{ \n',
    '    req(gsub("^Assay: ", "", input${prefix}s2ass1)) \n',
    '    if(gsub("^Assay: ", "", input${prefix}s2ass1) == "Cell Information"){{ \n',
    '      res <- {prefix}conf$UI \n',
    '      resDef <- {prefix}def$meta1 \n',
    '      resLen <- length(res) \n',
    '    }} else {{ \n',
    '      res <- names({prefix}gene[[gsub("^Assay: ", "", input${prefix}s2ass1)]]) \n',
    '      resDef <- {prefix}def$gene1[[gsub("^Assay: ", "", input${prefix}s2ass1)]] \n',
    '      resLen <- 7 \n',
    '    }} \n',
    '    return(list(res,resDef,resLen))\n',
    '  }})\n',
    '  observeEvent(gsub("^Assay: ", "", input${prefix}s2ass1), {{\n',
    '    updateSelectizeInput(session, "{prefix}s2inp1", choices = getG{prefix}s2inp1()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}s2inp1()[[2]], options = list( \n',
    '                           maxOptions = getG{prefix}s2inp1()[[3]], create = TRUE, \n',
    '                           persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '  getG{prefix}s2inp2 <- reactive({{ \n',
    '    req(gsub("^Assay: ", "", input${prefix}s2ass2)) \n',
    '    if(gsub("^Assay: ", "", input${prefix}s2ass2) == "Cell Information"){{ \n',
    '      res <- {prefix}conf$UI \n',
    '      resDef <- {prefix}def$meta1 \n',
    '      resLen <- length(res) \n',
    '    }} else {{ \n',
    '      res <- names({prefix}gene[[gsub("^Assay: ", "", input${prefix}s2ass2)]]) \n',
    '      resDef <- {prefix}def$gene1[[gsub("^Assay: ", "", input${prefix}s2ass2)]] \n',
    '      resLen <- 7 \n',
    '    }} \n',
    '    return(list(res,resDef,resLen))\n',
    '  }})\n',
    '  observeEvent(gsub("^Assay: ", "", input${prefix}s2ass2), {{\n',
    '    updateSelectizeInput(session, "{prefix}s2inp2", choices = getG{prefix}s2inp2()[[1]], \n',
    '                         server = TRUE, selected = getG{prefix}s2inp2()[[2]], options = list( \n',
    '                           maxOptions = getG{prefix}s2inp2()[[3]], create = TRUE, \n',
    '                           persist = TRUE, render = I(optCrt))) \n',
    '  }})\n',
    '    output${prefix}s2sub1.ui <- renderUI({{ \n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}s2sub1]$fID, "\\\\|")[[1]] \n',
    '    checkboxGroupInput("{prefix}s2sub2", "Select which cells to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}s2sub1non, {{\n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}s2sub1]$fID, "\\\\|")[[1]]\n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}s2sub2", label = "Select which cells to show",\n',
    '                             choices = sub, selected = NULL, inline = TRUE)\n',
    '  }})\n',
    '  observeEvent(input${prefix}s2sub1all, {{\n',
    '    sub = strsplit({prefix}conf[UI == input${prefix}s2sub1]$fID, "\\\\|")[[1]]\n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}s2sub2", label = "Select which cells to show",\n',
    '                             choices = sub, selected = sub, inline = TRUE)\n',
    '  }})\n',
    '  \n',
    '  {prefix}s2oup1 <- reactive({{\n',
    '    sc2Dspat({prefix}conf, {prefix}meta, {prefix}image, input${prefix}s2alp, \n',
    '             input${prefix}s2inp1, "{prefix}assay_", {prefix}gene, input${prefix}s2ass1, \n',
    '             input${prefix}s2sub1, input${prefix}s2sub2, input${prefix}s2min1, input${prefix}s2max1, input${prefix}s2siz, \n',
    '             cList[[input${prefix}s2col1]], sList[input${prefix}s2fsz], input${prefix}s2lab1) \n',
    '  }})\n',
    '  output${prefix}s2oup1 <- renderPlot({{{prefix}s2oup1() + theme(legend.position = "none")}})\n',
    '  output${prefix}s2oup1.ui <- renderUI({{plotOutput("{prefix}s2oup1", height = pList[input${prefix}s2psz])}})\n',
    '  output${prefix}s2oup1.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}spatial_",input${prefix}s2inp1,".",input${prefix}s2oup1.f) }},\n',
    '    content = function(file) {{ ggsav(file, height = input${prefix}s2oup1.h, width = input${prefix}s2oup1.w, \n',
    '                                     plot = {prefix}s2oup1() + theme(legend.position = "none"))\n',
    '  }})\n',
    '  output${prefix}s2oup3 <- renderPlot({{grid.newpage(); grid.draw(g_legend({prefix}s2oup1()))}})\n',
    '  output${prefix}s2oup3.ui <- renderUI({{plotOutput("{prefix}s2oup3", height = 72*convertHeight(\n',
    '    grobHeight(g_legend({prefix}s2oup1())), unitTo="in", valueOnly=TRUE) + 50)}})\n',
    '  output${prefix}s2oup3.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}spatial_",input${prefix}s2inp1,"_leg.",input${prefix}s2oup3.f) }},\n',
    '    content = function(file) {{ \n',
    '      grid.newpage(); grid.draw(g_legend({prefix}s2oup1()))\n',
    '      ggsav(file, height = input${prefix}s2oup3.h, width = input${prefix}s2oup3.w, plot = grid.grab())\n',
    '  }})\n',
    '  {prefix}s2oup2 <- reactive({{\n',
    '    sc2Dspat({prefix}conf, {prefix}meta, {prefix}image, input${prefix}s2alp, \n',
    '             input${prefix}s2inp2, "{prefix}assay_", {prefix}gene, input${prefix}s2ass2, \n',
    '             input${prefix}s2sub1, input${prefix}s2sub2, input${prefix}s2min2, input${prefix}s2max2, input${prefix}s2siz, \n',
    '             cList[[input${prefix}s2col2]], sList[input${prefix}s2fsz], input${prefix}s2lab2) \n',
    '  }})\n',
    '  output${prefix}s2oup2 <- renderPlot({{{prefix}s2oup2() + theme(legend.position = "none")}})\n',
    '  output${prefix}s2oup2.ui <- renderUI({{plotOutput("{prefix}s2oup2", height = pList[input${prefix}s2psz])}})\n',
    '  output${prefix}s2oup2.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}spatial_",input${prefix}s2inp2,".",input${prefix}s2oup4.f) }},\n',
    '    content = function(file) {{ ggsav(file, height = input${prefix}s2oup2.h, width = input${prefix}s2oup2.w, \n',
    '                                     plot = {prefix}s2oup2() + theme(legend.position = "none"))\n',
    '  }})\n',
    '  output${prefix}s2oup4 <- renderPlot({{grid.newpage(); grid.draw(g_legend({prefix}s2oup2()))}})\n',
    '  output${prefix}s2oup4.ui <- renderUI({{plotOutput("{prefix}s2oup4", height = 72*convertHeight(\n',
    '    grobHeight(g_legend({prefix}s2oup2())), unitTo="in", valueOnly=TRUE) + 50)}})\n',
    '  output${prefix}s2oup4.dl <- downloadHandler(\n',
    '    filename = function() {{ paste0("{prefix}spatial_",input${prefix}s2inp2,"_leg.",input${prefix}s2oup4.f) }},\n',
    '    content = function(file) {{ \n',
    '      grid.newpage(); grid.draw(g_legend({prefix}s2oup2()))\n',
    '      ggsav(file, height = input${prefix}s2oup4.h, width = input${prefix}s2oup4.w, plot = grid.grab())\n',
    '  }})\n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVmainT1
#' @export wrSVmainT1
#'
wrSVmainT1 <- function(prefix) {
  glue::glue(
    '  ### Functions for tab T1 \n',
    '  updateSelectizeInput(session, "{prefix}t1gene", choices = {prefix}bw$chrGene$gene, server = TRUE,  \n',
    '                       selected = {prefix}bw$chrGene$gene[42], options = list( \n',
    '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
    '  observe({{ \n',
    '    files <- c("no file selected", list.files(path = "./bed_files", pattern = "*.bed|*.tsv")) \n',
    '    updateSelectInput(session, "{prefix}t1bed", choices = files, selected = "no file selected") \n',
    '  }}) \n',
    '  output${prefix}t1sub1.ui <- renderUI({{ \n',
    '    sub = dir(paste0("{prefix}bw_", input${prefix}t1grp), full.names = FALSE, pattern = "*.bw") \n',
    '    sub = gsub(".bw$", "", sub) \n',
    '    checkboxGroupInput("{prefix}t1sub2", "Select which tracks to show", inline = TRUE, \n',
    '                       choices = sub, selected = sub) \n',
    '  }}) \n',
    '  observeEvent(input${prefix}t1sub1non, {{  \n',
    '    sub = dir(paste0("{prefix}bw_", input${prefix}t1grp), full.names = FALSE, pattern = "*.bw") \n',
    '    sub = gsub(".bw$", "", sub) \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}t1sub2", label = "Select which tracks to show",  \n',
    '                             choices = sub, selected = NULL, inline = TRUE)  \n',
    '  }})  \n',
    '  observeEvent(input${prefix}t1sub1all, {{  \n',
    '    sub = dir(paste0("{prefix}bw_", input${prefix}t1grp), full.names = FALSE, pattern = "*.bw") \n',
    '    sub = gsub(".bw$", "", sub) \n',
    '    updateCheckboxGroupInput(session, inputId = "{prefix}t1sub2", label = "Select which tracks to show",  \n',
    '                             choices = sub, selected = sub, inline = TRUE)  \n',
    '  }})  \n',
    '  \n',
    '  {prefix}bwGroup <- reactive({{ \n',
    '    req(input${prefix}t1grp) \n',
    '    bigWigs <- dir(paste0("{prefix}bw_", input${prefix}t1grp), full.names = TRUE) \n',
    '    bw_samples <- read_coldata(bws = bigWigs, build = {prefix}bw$genome) \n',
    '    if(!is.null(input${prefix}t1sub2[[1]])) {{ \n',
    '      tmp <- bw_samples[bw_files %in% paste0( \n',
    '        "{prefix}bw_", input${prefix}t1grp, "/", input${prefix}t1sub2, ".bw"), ] \n',
    '      if(nrow(tmp) > 0){{bw_samples <- tmp}} \n',
    '    }} \n',
    '    return(bw_samples) \n',
    '  }}) \n',
    '  {prefix}t1tr <- eventReactive(input${prefix}t1update, {{ \n',
    '    # Process track color \n',
    '    tmp = strsplit({prefix}conf[UI == input${prefix}t1grp]$fCL, "\\\\|")[[1]] \n',
    '    names(tmp) = strsplit({prefix}conf[UI == input${prefix}t1grp]$fID, "\\\\|")[[1]] \n',
    '    if(!is.null(input${prefix}t1sub2)){{ \n',
    '      track_col = tmp[input${prefix}t1sub2] \n',
    '    }} else {{ \n',
    '      track_col = tmp \n',
    '    }} \n',
    '    if(is.na(track_col[[1]])){{track_col = "grey"}} \n',
    '    if(input${prefix}t1col == FALSE){{track_col = "grey"}} \n',
    '    # Collect inputs \n',
    '    inp <- list(gene = input${prefix}t1gene, \n',
    '                chr  = input${prefix}t1chr, \n',
    '                srt  = input${prefix}t1srt, \n',
    '                end  = input${prefix}t1end, \n',
    '                bed  = input${prefix}t1bed, \n',
    '                extL = input${prefix}t1extL, \n',
    '                extR = input${prefix}t1extR, \n',
    '                col  = track_col, \n',
    '                bw_colData = {prefix}bwGroup()) \n',
    '    if (input${prefix}t1sel == "Region"){{inp$gene = NULL}} \n',
    '    if (input${prefix}t1bed == "no file selected"){{inp$bed = NULL}} \n',
    '    inp \n',
    '  }}, ignoreNULL = FALSE) \n',
    '  output${prefix}t1track <- renderPlot({{ \n',
    '    inp <- {prefix}t1tr() \n',
    '    scTrackplot("./", inp$bw_colData, inp$gene, {prefix}bw, inp$chr, inp$srt, inp$end, \n',
    '                inp$extL, inp$extR, inp$bed, "{prefix}bw.gtf.gz", binSize = 100,  \n',
    '                input${prefix}t1dgt, input${prefix}t1asg, input${prefix}t1gth,  \n',
    '                input${prefix}t1ath, input${prefix}t1sha, input${prefix}t1gsz,  \n',
    '                input${prefix}t1tnp, input${prefix}t1ctx, inp$col) \n',
    '  }}) \n',
    '  output${prefix}t1track.ui <- renderUI({{ \n',
    '    plotOutput("{prefix}t1track",  \n',
    '               height = paste0(input${prefix}t1track.h, "px"),  \n',
    '               width =  paste0(input${prefix}t1track.w, "px")) \n',
    '  }}) \n',
    '\n',
    '\n',
    '\n'
  )
}

#' Write code for server.R
#'
#' @rdname wrSVpost
#' @export wrSVpost
#'
wrSVpost <- function() {
  glue::glue(
    '}}) \n',
    '\n',
    '\n',
    '\n'
  )
}


