#' Write code for ui.R
#'
#' @rdname wrUIlib
#' @export wrUIlib
#'
wrUIlib <- function() {
  glue::glue(
    'library(shiny) \n',
    'library(shinyhelper) \n',
    'library(data.table) \n',
    'library(Matrix) \n',
    'library(DT) \n',
    'library(magrittr) \n',
    '\n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUIload
#' @export wrUIload
#'
wrUIload <- function(prefix) {
  glue::glue(
    '{prefix}conf = readRDS("./{prefix}conf.rds")\n',
    '{prefix}def  = readRDS("./{prefix}def.rds")\n',
    '\n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUIloadT1
#' @export wrUIloadT1
#'
wrUIloadT1 <- function(prefix) {
  glue::glue(
    '{prefix}bw    = readRDS("./{prefix}bw.rds")\n',
    '{prefix}bwGrp = tstrsplit(dir(pattern = "{prefix}bw_"), "_")[[2]]\n',
    '\n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUIpre
#' @export wrUIpre
#'
wrUIpre <- function(title, ganalytics) {
  if(!is.na(ganalytics)){
    ga = 'tags$head(includeHTML(("google-analytics.html"))),'
  } else {
    ga = ''
  }
  glue::glue(
    '\n',
    '\n',
    '\n',
    '### Start server code \n',
    'shinyUI(fluidPage( \n',
    '  ### HTML formatting of error messages \n',
    '  {ga} \n',
    '  tags$head(tags$style(HTML(".shiny-output-error-validation {{color: red; font-weight: bold;}}"))), \n',
    '  list(tags$style(HTML(".navbar-default .navbar-nav {{ font-weight: bold; font-size: 16px; }}"))), \n',
    '  \n',
    '  ### Page title \n',
    '  titlePanel("{title}"), \n',
    '  navbarPage( \n',
    '    NULL, \n',
    '\n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainA1
#' @export wrUImainA1
#'
wrUImainA1 <- function(prefix, ptsiz) {
  glue::glue(
    '  ### Tab1.a1: Zoom-enable Dimred \n',
    '  tabPanel( \n',
    '    HTML("Zoom-enable Dimred"), \n',
    '    h4("Zoom-enable reduced dimensions overlaid with cell info or assay expression"), \n',
    '    "In this tab, users can visualise either cell information and gene expression ",  \n',
    '    "in a zoom-enabled low-dimensional representions plots.", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, fluidRow( \n',
    '          column( \n',
    '            12, selectInput("{prefix}a1dr", "Reduction:", choices = {prefix}def$dimrd, \n',
    '                            selected = {prefix}def$dimrd[1])) \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        3, actionButton("{prefix}a1togL", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}a1togL % 2 == 1", \n',
    '          selectInput("{prefix}a1sub1", "Cell information to subset:", \n',
    '                      choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                      selected = {prefix}def$grp1), \n',
    '          uiOutput("{prefix}a1sub1.ui"), \n',
    '          actionButton("{prefix}a1sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}a1sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        6, actionButton("{prefix}a1tog0", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}a1tog0 % 2 == 1", \n',
    '          fluidRow( \n',
    '            column( \n',
    '              6, sliderInput("{prefix}a1siz", "Point size:", \n',
    '                             min = 0, max = 4, value = {ptsiz}, step = 0.25), \n',
    '              radioButtons("{prefix}a1psz", "Plot size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE), \n',
    '              radioButtons("{prefix}a1fsz", "Font size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE) \n',
    '            ), \n',
    '            column( \n',
    '              6, radioButtons("{prefix}a1asp", "Aspect ratio:", \n',
    '                              choices = c("Square", "Fixed", "Free"), \n',
    '                              selected = "Square", inline = TRUE), \n',
    '              checkboxInput("{prefix}a1txt", "Show axis text", value = FALSE) \n',
    '            ) \n',
    '          ) \n',
    '        ) \n',
    '      )  # End of column (6 space) \n',
    '    ),   # End of fluidRow (4 space) \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, style="border-right: 2px solid black", h4("Information to plot"),\n',
    '        selectInput("{prefix}a1ass1", "Data type to colour plot:",\n',
    '                    choices = c("Cell Information", paste0("Assay: ", {prefix}def$assay)), \n',
    '                    selected = "Cell Information"), \n',
    '        selectInput("{prefix}a1inp1", "Cell Info / Feature Name:", choices = NULL) %>% \n',
    '          helper(type = "inline", size = "m", fade = TRUE,\n',
    '                 title = "Cell Info / Gene to colour cells by",\n',
    '                 content = c("Select cell info / feature to colour cells",\n',
    '                             "- Categorical covariates have a fixed colour palette",\n',
    '                             paste0("- Continuous covariates / gene expression are coloured ",\n',
    '                                    "in a Blue-Yellow-Red colour scheme, which can be ",\n',
    '                                    "changed in the plot controls"))),\n',
    '        strong("Draw box on plot below to zoom: "), \n',
    '        plotOutput("{prefix}a1oup1.br", height = "400px",\n',
    '                   brush = brushOpts(id = "{prefix}a1inp1.br", resetOnNew = TRUE)), \n',
    '        actionButton("{prefix}a1tog1", "Toggle plot controls"), \n',
    '        conditionalPanel(\n',
    '          condition = "input.{prefix}a1tog1 % 2 == 1",\n',
    '          radioButtons("{prefix}a1col1", "Colour (Continuous data):",\n',
    '                       choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"),\n',
    '                       selected = "Blue-Yellow-Red"),\n',
    '          numericInput("{prefix}a1min1", "Min cutoff (q##):",\n',
    '                       min = 0, max = 50, value = 0, step = 1),\n',
    '          numericInput("{prefix}a1max1", "Max cutoff (q##):",\n',
    '                       min = 50, max = 100, value = 100, step = 1),\n',
    '          radioButtons("{prefix}a1ord1", "Plot order:",\n',
    '                       choices = c("Max-1st", "Min-1st", "Original", "Random"),\n',
    '                       selected = "Original", inline = TRUE),\n',
    '          checkboxInput("{prefix}a1lab1", "Show cell info labels", value = TRUE)\n',
    '        )\n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        6, style="border-right: 2px solid black", \n',
    '        fluidRow(column(12, uiOutput("{prefix}a1oup1.ui"))),\n',
    '        fluidRow(column(12, uiOutput("{prefix}a1oup3.ui"))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}a1oup1.dl", "Download Plot")),\n',
    '                 column(3, radioButtons("{prefix}a1oup1.f", "Plot format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}a1oup1.h", "Plot height:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)),\n',
    '                 column(3, numericInput("{prefix}a1oup1.w", "Plot width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}a1oup3.dl", "Download Legend")),\n',
    '                 column(3, radioButtons("{prefix}a1oup3.f", "Legend format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}a1oup3.h", "Legend height:",\n',
    '                                        min = 0.2, max = 10, value = 2, step = 0.1)),\n',
    '                 column(3, numericInput("{prefix}a1oup3.w", "Legend width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        br() \n',
    '      ), # End of column (6 space) \n',
    '      column(\n',
    '        3, h4("Cell numbers"), \n',
    '        numericInput("{prefix}a1splt", "Split continuous cell info into nBins:",\n',
    '                     min = 2, max = 10, value = 4, step = 1),\n',
    '        dataTableOutput("{prefix}a1.dt") \n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  , \n',
    '  \n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainA2
#' @export wrUImainA2
#'
wrUImainA2 <- function(prefix, ptsiz) {
  glue::glue(
    '  ### Tab1.a2: CellInfo vs AssayExpr on dimRed \n',
    '  tabPanel( \n',
    '    HTML("Side-by-side DimRed"), \n',
    '    h4("Cell information vs assay expression on reduced dimensions"), \n',
    '    "In this tab, users can visualise both cell information and gene ",  \n',
    '    "expression side-by-side on low-dimensional representions.", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, fluidRow(\n',
    '          column(\n',
    '            12, selectInput("{prefix}a2dr", "Reduction:", choices = {prefix}def$dimrd, \n',
    '                            selected = {prefix}def$dimrd[1]))\n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        3, actionButton("{prefix}a2togL", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}a2togL % 2 == 1", \n',
    '          selectInput("{prefix}a2sub1", "Cell information to subset:", \n',
    '                      choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                      selected = {prefix}def$grp1), \n',
    '          uiOutput("{prefix}a2sub1.ui"), \n',
    '          actionButton("{prefix}a2sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}a2sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        6, actionButton("{prefix}a2tog0", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}a2tog0 % 2 == 1", \n',
    '          fluidRow( \n',
    '            column( \n',
    '              6, sliderInput("{prefix}a2siz", "Point size:", \n',
    '                             min = 0, max = 4, value = {ptsiz}, step = 0.25), \n',
    '              radioButtons("{prefix}a2psz", "Plot size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE), \n',
    '              radioButtons("{prefix}a2fsz", "Font size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE) \n',
    '            ), \n',
    '            column( \n',
    '              6, radioButtons("{prefix}a2asp", "Aspect ratio:", \n',
    '                              choices = c("Square", "Fixed", "Free"), \n',
    '                              selected = "Square", inline = TRUE), \n',
    '              checkboxInput("{prefix}a2txt", "Show axis text", value = FALSE) \n',
    '            ) \n',
    '          ) \n',
    '        ) \n',
    '      )  # End of column (6 space) \n',
    '    ),   # End of fluidRow (4 space) \n',
    '    fluidRow( \n',
    '      column(\n',
    '        6, style="border-right: 2px solid black",\n',
    '        fluidRow(\n',
    '          column(\n',
    '            6, selectInput("{prefix}a2ass1", "Data type to colour plot:",\n',
    '                           choices = c("Cell Information", paste0("Assay: ", {prefix}def$assay)), \n',
    '                           selected = "Cell Information"), \n',
    '            selectInput("{prefix}a2inp1", "Cell Info / Feature Name:", choices = NULL) %>% \n',
    '              helper(type = "inline", size = "m", fade = TRUE,\n',
    '                     title = "Cell Info / Gene to colour cells by",\n',
    '                     content = c("Select cell info / feature to colour cells",\n',
    '                                 "- Categorical covariates have a fixed colour palette",\n',
    '                                 paste0("- Continuous covariates / gene expression are coloured ",\n',
    '                                        "in a Blue-Yellow-Red colour scheme, which can be ",\n',
    '                                        "changed in the plot controls")))\n',
    '          ),\n',
    '          column(\n',
    '            6, actionButton("{prefix}a2tog1", "Toggle plot controls"),\n',
    '            conditionalPanel(\n',
    '              condition = "input.{prefix}a2tog1 % 2 == 1",\n',
    '              radioButtons("{prefix}a2col1", "Colour (Continuous data):",\n',
    '                           choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"),\n',
    '                           selected = "Blue-Yellow-Red"),\n',
    '              numericInput("{prefix}a2min1", "Min cutoff (q##):",\n',
    '                           min = 0, max = 50, value = 0, step = 1),\n',
    '              numericInput("{prefix}a2max1", "Max cutoff (q##):",\n',
    '                           min = 50, max = 100, value = 100, step = 1),\n',
    '              radioButtons("{prefix}a2ord1", "Plot order:",\n',
    '                           choices = c("Max-1st", "Min-1st", "Original", "Random"),\n',
    '                           selected = "Original", inline = TRUE),\n',
    '              checkboxInput("{prefix}a2lab1", "Show cell info labels", value = TRUE)\n',
    '            )\n',
    '          )\n',
    '        ),\n',
    '        fluidRow(column(12, uiOutput("{prefix}a2oup1.ui"))),\n',
    '        fluidRow(column(12, uiOutput("{prefix}a2oup3.ui"))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}a2oup1.dl", "Download Plot")),\n',
    '                 column(3, radioButtons("{prefix}a2oup1.f", "Plot format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}a2oup1.h", "Plot height:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)),\n',
    '                 column(3, numericInput("{prefix}a2oup1.w", "Plot width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}a2oup3.dl", "Download Legend")),\n',
    '                column(3, radioButtons("{prefix}a2oup3.f", "Legend format:", \n',
    '                                       choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}a2oup3.h", "Legend height:",\n',
    '                                        min = 0.2, max = 10, value = 2, step = 0.1)),\n',
    '                 column(3, numericInput("{prefix}a2oup3.w", "Legend width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        br() \n',
    '      ), # End of column (6 space)\n',
    '      column( \n',
    '        6,\n',
    '        fluidRow( \n',
    '          column(\n',
    '            6, selectInput("{prefix}a2ass2", "Data type to colour plot:",\n',
    '                           choices = c("Cell Information", paste0("Assay: ", {prefix}def$assay)), \n',
    '                           selected = paste0("Assay: ", {prefix}def$assay[1])), \n',
    '            selectInput("{prefix}a2inp2", "Cell Info / Feature Name:", choices = NULL) %>% \n',
    '              helper(type = "inline", size = "m", fade = TRUE,\n',
    '                     title = "Cell Info / Gene to colour cells by",\n',
    '                     content = c("Select cell info / feature to colour cells",\n',
    '                                 "- Categorical covariates have a fixed colour palette",\n',
    '                                 paste0("- Continuous covariates / gene expression are coloured ",\n',
    '                                        "in a Blue-Yellow-Red colour scheme, which can be ",\n',
    '                                        "changed in the plot controls")))\n',
    '          ),\n',
    '          column(\n',
    '            6, actionButton("{prefix}a2tog2", "Toggle plot controls"),\n',
    '            conditionalPanel(\n',
    '              condition = "input.{prefix}a2tog2 % 2 == 1",\n',
    '              radioButtons("{prefix}a2col2", "Colour (Continuous data):",\n',
    '                           choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"),\n',
    '                           selected = "White-Red"),\n',
    '              numericInput("{prefix}a2min2", "Min cutoff (q##):",\n',
    '                           min = 0, max = 50, value = 0, step = 1),\n',
    '              numericInput("{prefix}a2max2", "Max cutoff (q##):",\n',
    '                           min = 50, max = 100, value = 100, step = 1),\n',
    '              radioButtons("{prefix}a2ord2", "Plot order:",\n',
    '                           choices = c("Max-1st", "Min-1st", "Original", "Random"),\n',
    '                           selected = "Max-1st", inline = TRUE),\n',
    '              checkboxInput("{prefix}a2lab2", "Show cell info labels", value = TRUE)\n',
    '            )\n',
    '          )\n',
    '        ) , \n',
    '        fluidRow(column(12, uiOutput("{prefix}a2oup2.ui"))),\n',
    '        fluidRow(column(12, uiOutput("{prefix}a2oup4.ui"))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}a2oup2.dl", "Download Plot")),\n',
    '                 column(3, radioButtons("{prefix}a2oup2.f", "Plot format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}a2oup2.h", "Plot height:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)),\n',
    '                 column(3, numericInput("{prefix}a2oup2.w", "Plot width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}a2oup4.dl", "Download Legend")),\n',
    '                 column(3, radioButtons("{prefix}a2oup4.f", "Legend format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}a2oup4.h", "Legend height:",\n',
    '                                        min = 0.2, max = 10, value = 2, step = 0.1)),\n',
    '                 column(3, numericInput("{prefix}a2oup4.w", "Legend width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        br()\n',
    '      )  # End of column (6 space) \n',
    '    ),   # End of fluidRow (4 space) \n',
    '    h4("Relationship between left-side and right-side cell info / feature"),\n',
    '    fluidRow(\n',
    '      column(\n',
    '        6, h4("Comparative plot"), \n',
    '        style="border-right: 2px solid black", \n',
    '        uiOutput("{prefix}a2oup5.ui"), \n',
    '        fluidRow(column(3, downloadButton("{prefix}a2oup5.dl", "Download Plot")),\n',
    '                 column(3, radioButtons("{prefix}a2oup5.f", "Plot format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}a2oup5.h", "Plot height:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)),\n',
    '                 column(3, numericInput("{prefix}a2oup5.w", "Plot width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)))\n',
    '      ),\n',
    '      column(\n',
    '        6, h4("Cell numbers / statistics"), \n',
    '        numericInput("{prefix}a2cut", "Cutoff for Expression:", value = 0), \n',
    '        dataTableOutput("{prefix}a2.dt") \n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  , \n',
    '  \n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainA3
#' @export wrUImainA3
#'
wrUImainA3 <- function(prefix, ptsiz) {
  glue::glue(
    '  ### Tab1.a3: Gene coexpression plot \n',
    '  tabPanel( \n',
    '    HTML("Gene coexpression"), \n',
    '    h4("Coexpression of two genes on reduced dimensions"), \n',
    '    "In this tab, users can visualise the coexpression of two genes ", \n',
    '    "on low-dimensional representions.", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, fluidRow( \n',
    '          column( \n',
    '            12, selectInput("{prefix}a3dr", "Reduction:", choices = {prefix}def$dimrd, \n',
    '                            selected = {prefix}def$dimrd[1])) \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        3, actionButton("{prefix}a3togL", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}a3togL % 2 == 1", \n',
    '          selectInput("{prefix}a3sub1", "Cell information to subset:", \n',
    '                      choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                      selected = {prefix}def$grp1), \n',
    '          uiOutput("{prefix}a3sub1.ui"), \n',
    '          actionButton("{prefix}a3sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}a3sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        6, actionButton("{prefix}a3tog0", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}a3tog0 % 2 == 1", \n',
    '          fluidRow( \n',
    '            column( \n',
    '              6, sliderInput("{prefix}a3siz", "Point size:", \n',
    '                             min = 0, max = 4, value = {ptsiz}, step = 0.25), \n',
    '              radioButtons("{prefix}a3psz", "Plot size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE), \n',
    '              radioButtons("{prefix}a3fsz", "Font size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE) \n',
    '            ), \n',
    '            column( \n',
    '              6, radioButtons("{prefix}a3asp", "Aspect ratio:", \n',
    '                              choices = c("Square", "Fixed", "Free"), \n',
    '                              selected = "Square", inline = TRUE), \n',
    '              checkboxInput("{prefix}a3txt", "Show axis text", value = FALSE) \n',
    '            ) \n',
    '          ) \n',
    '        ) \n',
    '      )  # End of column (6 space) \n',
    '    ),   # End of fluidRow (4 space) \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, style="border-right: 2px solid black", h4("Assay Expression"), \n',
    '        selectInput("{prefix}a3inp1", "Feature 1:", choices=NULL) %>%  \n',
    '          helper(type = "inline", size = "m", fade = TRUE, \n',
    '                title = "Feature expression to colour cells by", \n',
    '                content = c("Select gene to colour cells by gene expression", \n',
    '                           paste0("- Feature expression are coloured in a ", \n',
    '                                  "White-Red colour scheme which can be ", \n',
    '                                  "changed in the plot controls"))), \n',
    '        selectInput("{prefix}a3inp2", "Feature 2:", choices=NULL) %>% \n',
    '          helper(type = "inline", size = "m", fade = TRUE, \n',
    '                 title = "Feature expression to colour cells by", \n',
    '                 content = c("Select gene to colour cells by gene expression", \n',
    '                             paste0("- Feature expression are coloured in a ", \n',
    '                                    "White-Blue colour scheme which can be ", \n',
    '                                    "changed in the plot controls"))), \n',
    '        selectInput("{prefix}a3ass1", "Assay:", \n',
    '                    choices = {prefix}def$assay, \n',
    '                    selected = {prefix}def$assay[1]), \n',
    '        actionButton("{prefix}a3tog1", "Toggle plot controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}a3tog1 % 2 == 1", \n',
    '          radioButtons("{prefix}a3col1", "Colour:", \n',
    '                       choices = c("Red (Gene1); Blue (Gene2)", \n',
    '                                   "Orange (Gene1); Blue (Gene2)", \n',
    '                                   "Red (Gene1); Green (Gene2)", \n',
    '                                   "Green (Gene1); Blue (Gene2)"), \n',
    '                       selected = "Red (Gene1); Blue (Gene2)"), \n',
    '          numericInput("{prefix}a3min1", "Min cutoff gene 1 (q##):", \n',
    '                       min = 0, max = 50, value = 0, step = 1), \n',
    '          numericInput("{prefix}a3max1", "Max cutoff gene 1 (q##):", \n',
    '                       min = 50, max = 100, value = 100, step = 1), \n',
    '          numericInput("{prefix}a3min2", "Min cutoff gene 2 (q##):", \n',
    '                       min = 0, max = 50, value = 0, step = 1), \n',
    '          numericInput("{prefix}a3max2", "Max cutoff gene 2 (q##):", \n',
    '                       min = 50, max = 100, value = 100, step = 1), \n',
    '          radioButtons("{prefix}a3ord1", "Plot order:", \n',
    '                       choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
    '                       selected = "Max-1st", inline = TRUE) \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        6, style="border-right: 2px solid black", \n',
    '        uiOutput("{prefix}a3oup1.ui"), \n',
    '        fluidRow(column(3, downloadButton("{prefix}a3oup1.dl", "Download Plot")), \n',
    '                 column(3, radioButtons("{prefix}a3oup1.f", "Plot format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)), \n',
    '                 column(3, numericInput("{prefix}a3oup1.h", "Plot height:", \n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)), \n',
    '                 column(3, numericInput("{prefix}a3oup1.w", "Plot width:", \n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))), \n',
    '        br() \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        3, uiOutput("{prefix}a3oup2.ui"), \n',
    '        fluidRow(column(6, downloadButton("{prefix}a3oup2.dl", "Download Legend")), \n',
    '                 column(6, radioButtons("{prefix}a3oup2.f", "Legend format:", \n',
    '                        choices = c("png", "pdf"), selected = "png", inline = TRUE))), \n',
    '        br(), h4("Cell numbers"), \n',
    '        dataTableOutput("{prefix}a3.dt") \n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  , \n',
    '  \n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainB1
#' @export wrUImainB1
#'
wrUImainB1 <- function(prefix, ptsiz) {
  glue::glue(
    '  ### Tab1.b1: violinplot / boxplot \n',
    '  tabPanel( \n',
    '    HTML("Violinplot / Boxplot"),  \n',
    '    h4("Cell information / assay expression violin plot / box plot"), \n',
    '    "In this tab, users can visualise the assay expression or continuous cell information ",  \n',
    '    "(e.g. Number of UMIs / module score) across groups of cells (e.g. libary / clusters).", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, style="border-right: 2px solid black", \n',
    '        selectInput("{prefix}b1inp1", "Cell information (X-axis):", \n',
    '                    choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                    selected = {prefix}def$grp1) %>%  \n',
    '          helper(type = "inline", size = "m", fade = TRUE, \n',
    '                 title = "Cell information to group cells by",  \n',
    '                 content = c("Select categorical cell information to group cells by",  \n',
    '                             "- Single cells are grouped by this categorical covariate",  \n',
    '                             "- Plotted as the X-axis of the violin plot / box plot")),  \n',
    '        selectInput("{prefix}b1inp2", "Cell Info / Feature name (Y-axis):", choices=NULL) %>%  \n',
    '          helper(type = "inline", size = "m", fade = TRUE, \n',
    '                 title = "Cell Info / Gene to plot", \n',
    '                 content = c("Select cell info / feature to plot on Y-axis", \n',
    '                             "- Can be continuous cell information (e.g. nUMIs / scores)", \n',
    '                             "- Can also be feature expression")), \n',
    '        selectInput("{prefix}b1ass1", "Data type for Y-axis:", \n',
    '                    choices = c("Cell Information", paste0("Assay: ", {prefix}def$assay)), \n',
    '                    selected = "Cell Information"), \n',
    '        radioButtons("{prefix}b1typ", "Plot type:", \n',
    '                     choices = c("violin", "boxplot"), \n',
    '                     selected = "violin", inline = TRUE), \n',
    '        checkboxInput("{prefix}b1pts", "Show data points", value = FALSE), \n',
    '        actionButton("{prefix}b1togT", "Toggle to perform stats test(s)"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}b1togT % 2 == 1",  \n',
    '          radioButtons("{prefix}b1stg", "Perform global test:",  \n',
    '                       choices = c("none","kruskal.test","anova"),  \n',
    '                       selected = "none"), \n',
    '          radioButtons("{prefix}b1stp1", "Perform pairwise test:",  \n',
    '                       choices = c("none","wilcox.test","t.test"),  \n',
    '                       selected = "none"), \n',
    '          textAreaInput("{prefix}b1stp2", HTML("Pairwise comparisons to make <br />  \n',
    '                                               (Separate each pair by new line and <br />  \n',
    '                                               within each pair using , or ;):"),  \n',
    '                        height = "150px", value = "none") \n',
    '        ), br(), br(),  \n',
    '        actionButton("{prefix}b1togL", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}b1togL % 2 == 1", \n',
    '          selectInput("{prefix}b1sub1", "Cell information to subset:", \n',
    '                      choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                      selected = {prefix}def$grp1), \n',
    '          uiOutput("{prefix}b1sub1.ui"), \n',
    '          actionButton("{prefix}b1sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}b1sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ), br(), br(), \n',
    '        actionButton("{prefix}b1tog", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}b1tog % 2 == 1", \n',
    '          sliderInput("{prefix}b1siz", "Data point size:",  \n',
    '                      min = 0, max = 4, value = {ptsiz}, step = 0.25),  \n',
    '          radioButtons("{prefix}b1psz", "Plot size:", \n',
    '                       choices = c("Small", "Medium", "Large"), \n',
    '                       selected = "Medium", inline = TRUE), \n',
    '          radioButtons("{prefix}b1fsz", "Font size:", \n',
    '                       choices = c("Small", "Medium", "Large"), \n',
    '                       selected = "Medium", inline = TRUE), \n',
    '          checkboxInput("{prefix}b1noi", "Add noise to assay expr", value = TRUE)) \n',
    '      ), # End of column (6 space) \n',
    '      column(9, uiOutput("{prefix}b1oup.ui"), \n',
    '      fluidRow(column(2, downloadButton("{prefix}b1oup.dl", "Download Plot")), \n',
    '               column(2, radioButtons("{prefix}b1oup.f", "Plot format:", \n',
    '                                      choices = c("png", "pdf"), selected = "png", inline = TRUE)), \n',
    '               column(2, numericInput("{prefix}b1oup.h", "Plot height:", \n',
    '                                      min = 4, max = 20, value = 8, step = 0.5)), \n',
    '               column(2, numericInput("{prefix}b1oup.w", "Plot width:", \n',
    '                                      min = 4, max = 20, value = 10, step = 0.5))), \n',
    '      br() \n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  , \n',
    '  \n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainB2
#' @export wrUImainB2
#'
wrUImainB2 <- function(prefix) {
  glue::glue(
    '  ### Tab1.b2: Proportion plot \n',
    '  tabPanel( \n',
    '    HTML("Proportion plot"), \n',
    '    h4("Proportion / cell numbers across different cell information"), \n',
    '    "In this tab, users can visualise the composition of single cells based on one discrete ", \n',
    '    "cell information across another discrete cell information. ",  \n',
    '    "Usage examples include the library or cellcycle composition across clusters.", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, style="border-right: 2px solid black", \n',
    '        selectInput("{prefix}b2inp1", "Cell information to plot (X-axis):", \n',
    '                    choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                    selected = {prefix}def$grp2) %>%  \n',
    '          helper(type = "inline", size = "m", fade = TRUE, \n',
    '                 title = "Cell information to plot cells by",  \n',
    '                 content = c("Select categorical cell information to plot cells by", \n',
    '                             "- Plotted as the X-axis of the proportion plot")), \n',
    '        selectInput("{prefix}b2inp2", "Cell information to group / colour by:", \n',
    '                    choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                    selected = {prefix}def$grp1) %>%  \n',
    '          helper(type = "inline", size = "m", fade = TRUE, \n',
    '                 title = "Cell information to group / colour cells by", \n',
    '                 content = c("Select categorical cell information to group / colour cells by", \n',
    '                             "- Proportion / cell numbers are shown in different colours")), \n',
    '        radioButtons("{prefix}b2typ", "Plot value:", \n',
    '                     choices = c("Proportion", "CellNumbers"), \n',
    '                     selected = "Proportion", inline = TRUE), \n',
    '        checkboxInput("{prefix}b2flp", "Flip X/Y", value = FALSE), \n',
    '        selectInput("{prefix}b2ord1", "Reorder X-axis by which group:", choices=NULL), \n',
    '        radioButtons("{prefix}b2ord2", "Reorder in which order:", \n',
    '                     choices = c("Decreasing", "Increasing"), \n',
    '                     selected = "Decreasing", inline = TRUE), \n',
    '        actionButton("{prefix}b2togL", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}b2togL % 2 == 1", \n',
    '          selectInput("{prefix}b2sub1", "Cell information to subset:", \n',
    '                      choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                      selected = {prefix}def$grp1), \n',
    '          uiOutput("{prefix}b2sub1.ui"), \n',
    '          actionButton("{prefix}b2sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}b2sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ), br(), br(), \n',
    '        actionButton("{prefix}b2tog", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}b2tog % 2 == 1", \n',
    '          radioButtons("{prefix}b2psz", "Plot size:", \n',
    '                       choices = c("Small", "Medium", "Large"), \n',
    '                       selected = "Medium", inline = TRUE), \n',
    '          radioButtons("{prefix}b2fsz", "Font size:", \n',
    '                       choices = c("Small", "Medium", "Large"), \n',
    '                       selected = "Medium", inline = TRUE)) \n',
    '      ), # End of column (6 space) \n',
    '      column(9, uiOutput("{prefix}b2oup.ui"), \n',
    '             fluidRow(column(2, downloadButton("{prefix}b2oup.dl", "Download Plot")), \n',
    '                      column(2, radioButtons("{prefix}b2oup.f", "Plot format:", \n',
    '                                             choices = c("png", "pdf"), selected = "png", inline = TRUE)), \n',
    '                      column(2, numericInput("{prefix}b2oup.h", "Plot height:", \n',
    '                                             min = 4, max = 20, value = 8, step = 0.5)), \n',
    '                      column(2, numericInput("{prefix}b2oup.w", "Plot width:", \n',
    '                                             min = 4, max = 20, value = 10, step = 0.5))), \n',
    '             br() \n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  , \n',
    '  \n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainB3
#' @export wrUImainB3
#'
wrUImainB3 <- function(prefix) {
  glue::glue(
    '  ### Tab1.b3: Bubbleplot / Heatmap \n',
    '  tabPanel( \n',
    '    HTML("Bubbleplot / Heatmap"), \n',
    '    h4("Gene expression bubbleplot / heatmap"), \n',
    '    "In this tab, users can visualise the gene expression patterns of ", \n',
    '    "multiple genes grouped by categorical cell information (e.g. library / cluster).", br(), \n',
    '    "The normalised expression are averaged, log-transformed and then plotted.", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, style="border-right: 2px solid black", \n',
    '        textAreaInput("{prefix}b3inp", HTML("List of gene names <br /> \n',
    '                                          (Max 50 genes, separated <br /> \n',
    '                                           by , or ; or newline):"), \n',
    '                      height = "250px", \n',
    '                      value = paste0({prefix}def$genes[[1]], collapse = ", ")) %>% \n',
    '          helper(type = "inline", size = "m", fade = TRUE, \n',
    '                 title = "List of genes to plot on bubbleplot / heatmap", \n',
    '                 content = c("Input genes to plot", \n',
    '                             "- Maximum 50 genes (due to ploting space limitations)", \n',
    '                             "- Genes should be separated by comma, semicolon or newline")), \n',
    '        selectInput("{prefix}b3ass1", "Assay:", \n',
    '                    choices = {prefix}def$assay, \n',
    '                    selected = {prefix}def$assay[1]), \n',
    '        selectInput("{prefix}b3grp", "Group by:", \n',
    '                    choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                    selected = {prefix}conf[grp == TRUE]$UI[1]) %>% \n',
    '          helper(type = "inline", size = "m", fade = TRUE, \n',
    '                 title = "Cell information to group cells by", \n',
    '                 content = c("Select categorical cell information to group cells by", \n',
    '                             "- Single cells are grouped by this categorical covariate", \n',
    '                             "- Plotted as the X-axis of the bubbleplot / heatmap")), \n',
    '        radioButtons("{prefix}b3plt", "Plot type:", \n',
    '                     choices = c("Bubbleplot", "Heatmap"), \n',
    '                     selected = "Bubbleplot", inline = TRUE), \n',
    '        checkboxInput("{prefix}b3scl", "Scale gene expression", value = TRUE), \n',
    '        checkboxInput("{prefix}b3row", "Cluster rows (genes)", value = TRUE), \n',
    '        checkboxInput("{prefix}b3col", "Cluster columns (samples)", value = FALSE), \n',
    '        sliderInput("{prefix}b3max", "Scale.max:", min=1, max=11, value=3, step=0.1), \n',
    '        checkboxInput("{prefix}b3exp", "expm1 before averaging (check for log-transformed data)",  \n',
    '                      value = TRUE), \n',
    '        br(), \n',
    '        actionButton("{prefix}b3togL", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}b3togL % 2 == 1", \n',
    '          selectInput("{prefix}b3sub1", "Cell information to subset:", \n',
    '                      choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                      selected = {prefix}def$grp1), \n',
    '          uiOutput("{prefix}b3sub1.ui"), \n',
    '          actionButton("{prefix}b3sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}b3sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ), br(), br(), \n',
    '        actionButton("{prefix}b3tog", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}b3tog % 2 == 1", \n',
    '          radioButtons("{prefix}b3cols", "Colour scheme:", \n',
    '                       choices = c("White-Red", "Blue-Yellow-Red", \n',
    '                                   "Yellow-Green-Purple"), \n',
    '                       selected = "Blue-Yellow-Red"), \n',
    '          radioButtons("{prefix}b3psz", "Plot size:", \n',
    '                       choices = c("Small", "Medium", "Large"), \n',
    '                       selected = "Medium", inline = TRUE), \n',
    '          radioButtons("{prefix}b3fsz", "Font size:", \n',
    '                       choices = c("Small", "Medium", "Large"), \n',
    '                       selected = "Medium", inline = TRUE)) \n',
    '      ), # End of column (6 space) \n',
    '      column(9, h4(htmlOutput("{prefix}b3oupTxt")), \n',
    '             uiOutput("{prefix}b3oup.ui"), \n',
    '             fluidRow(column(2, downloadButton("{prefix}b3oup.dl", "Download Plot")), \n',
    '                      column(2, radioButtons("{prefix}b3oup.f", "Plot format:", \n',
    '                                             choices = c("png", "pdf"), selected = "png", inline = TRUE)), \n',
    '                      column(2, numericInput("{prefix}b3oup.h", "Plot height:", \n',
    '                                             min = 4, max = 20, value = 10, step = 0.5)), \n',
    '                      column(2, numericInput("{prefix}b3oup.w", "Plot width:", \n',
    '                                             min = 4, max = 20, value = 10, step = 0.5))), \n',
    '             br() \n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  \n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainS1
#' @export wrUImainS1
#'
wrUImainS1 <- function(prefix, ptsiz) {
  glue::glue(
    '  ### Tab1.s1: Zoom-enable spatial \n',
    '  tabPanel( \n',
    '    HTML("Zoom-enable Spatial"), \n',
    '    h4("Zoom-enable spatial plot overlaid with cell info or assay expression"), \n',
    '    "In this tab, users can visualise either cell information and gene expression ",  \n',
    '    "in a zoom-enabled spatial plot.", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, fluidRow(\n',
    '          column(\n',
    '            12, sliderInput("{prefix}s1alp", "Dot transparency: ", \n',
    '                            min = 0, max = 1, value = c(0.5,1), step = 0.05))\n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        3, actionButton("{prefix}s1togL", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}s1togL % 2 == 1", \n',
    '          selectInput("{prefix}s1sub1", "Cell information to subset:", \n',
    '                      choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                      selected = {prefix}def$grp1), \n',
    '          uiOutput("{prefix}s1sub1.ui"), \n',
    '          actionButton("{prefix}s1sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}s1sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        6, actionButton("{prefix}s1tog0", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}s1tog0 % 2 == 1", \n',
    '          fluidRow( \n',
    '            column( \n',
    '              6, sliderInput("{prefix}s1siz", "Point size:", \n',
    '                             min = 0, max = 4, value = {ptsiz}, step = 0.25), \n',
    '              radioButtons("{prefix}s1psz", "Plot size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE), \n',
    '              radioButtons("{prefix}s1fsz", "Font size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE) \n',
    '            ) \n',
    '          ) \n',
    '        ) \n',
    '      )  # End of column (6 space) \n',
    '    ),   # End of fluidRow (4 space) \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, style="border-right: 2px solid black", h4("Information to plot"),\n',
    '        selectInput("{prefix}s1ass1", "Data type to colour plot:",\n',
    '                    choices = c("Cell Information", paste0("Assay: ", {prefix}def$assay)), \n',
    '                    selected = "Cell Information"), \n',
    '        selectInput("{prefix}s1inp1", "Cell Info / Feature Name:", choices = NULL) %>% \n',
    '          helper(type = "inline", size = "m", fade = TRUE,\n',
    '                 title = "Cell Info / Gene to colour cells by",\n',
    '                 content = c("Select cell info / feature to colour cells",\n',
    '                             "- Categorical covariates have a fixed colour palette",\n',
    '                             paste0("- Continuous covariates / gene expression are coloured ",\n',
    '                                    "in a Blue-Yellow-Red colour scheme, which can be ",\n',
    '                                    "changed in the plot controls"))),\n',
    '        strong("Draw box on plot below to zoom: "), \n',
    '        plotOutput("{prefix}s1oup1.br", height = "400px",\n',
    '                   brush = brushOpts(id = "{prefix}s1inp1.br", resetOnNew = TRUE)), \n',
    '        actionButton("{prefix}s1tog1", "Toggle plot controls"), \n',
    '        conditionalPanel(\n',
    '          condition = "input.{prefix}s1tog1 % 2 == 1",\n',
    '          radioButtons("{prefix}s1col1", "Colour (Continuous data):",\n',
    '                       choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"),\n',
    '                       selected = "Blue-Yellow-Red"),\n',
    '          numericInput("{prefix}s1min1", "Min cutoff (q##):",\n',
    '                       min = 0, max = 50, value = 0, step = 1),\n',
    '          numericInput("{prefix}s1max1", "Max cutoff (q##):",\n',
    '                       min = 50, max = 100, value = 100, step = 1),\n',
    '          checkboxInput("{prefix}s1lab1", "Show cell info labels", value = TRUE)\n',
    '        )\n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        6, style="border-right: 2px solid black", \n',
    '        fluidRow(column(12, uiOutput("{prefix}s1oup1.ui"))),\n',
    '        fluidRow(column(12, uiOutput("{prefix}s1oup3.ui"))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}s1oup1.dl", "Download Plot")),\n',
    '                 column(3, radioButtons("{prefix}s1oup1.f", "Plot format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}s1oup1.h", "Plot height:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)),\n',
    '                 column(3, numericInput("{prefix}s1oup1.w", "Plot width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}s1oup3.dl", "Download Legend")),\n',
    '                 column(3, radioButtons("{prefix}s1oup3.f", "Legend format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}s1oup3.h", "Legend height:",\n',
    '                                        min = 0.2, max = 10, value = 2, step = 0.1)),\n',
    '                 column(3, numericInput("{prefix}s1oup3.w", "Legend width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        br() \n',
    '      ), # End of column (6 space) \n',
    '      column(\n',
    '        3, h4("Cell numbers"), \n',
    '        numericInput("{prefix}s1splt", "Split continuous cell info into nBins:",\n',
    '                     min = 2, max = 10, value = 4, step = 1),\n',
    '        dataTableOutput("{prefix}s1.dt") \n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  , \n',
    '  \n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainS2
#' @export wrUImainS2
#'
wrUImainS2 <- function(prefix, ptsiz) {
  glue::glue(
    '  ### Tab1.s2: CellInfo vs AssayExpr on spatial \n',
    '  tabPanel( \n',
    '    HTML("Side-by-side Spatial"), \n',
    '    h4("Cell information vs assay expression on spatial coordinates"), \n',
    '    "In this tab, users can visualise both cell information and gene ",  \n',
    '    "expression side-by-side on spatial plot.", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, fluidRow(\n',
    '          column(\n',
    '            12, sliderInput("{prefix}s2alp", "Dot transparency: ", \n',
    '                            min = 0, max = 1, value = c(0.5,1), step = 0.05)) \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        3, actionButton("{prefix}s2togL", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}s2togL % 2 == 1", \n',
    '          selectInput("{prefix}s2sub1", "Cell information to subset:", \n',
    '                      choices = {prefix}conf[grp == TRUE]$UI, \n',
    '                      selected = {prefix}def$grp1), \n',
    '          uiOutput("{prefix}s2sub1.ui"), \n',
    '          actionButton("{prefix}s2sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}s2sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ) \n',
    '      ), # End of column (6 space) \n',
    '      column( \n',
    '        6, actionButton("{prefix}s2tog0", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}s2tog0 % 2 == 1", \n',
    '          fluidRow( \n',
    '            column( \n',
    '              6, sliderInput("{prefix}s2siz", "Point size:", \n',
    '                             min = 0, max = 5, value = {ptsiz}, step = 0.25), \n',
    '              radioButtons("{prefix}s2psz", "Plot size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE), \n',
    '              radioButtons("{prefix}s2fsz", "Font size:", \n',
    '                           choices = c("Small", "Medium", "Large"), \n',
    '                           selected = "Medium", inline = TRUE) \n',
    '            ) \n',
    '          ) \n',
    '        ) \n',
    '      )  # End of column (6 space) \n',
    '    ),   # End of fluidRow (4 space) \n',
    '    fluidRow( \n',
    '      column(\n',
    '        6, style="border-right: 2px solid black",\n',
    '        fluidRow(\n',
    '          column(\n',
    '            6, selectInput("{prefix}s2ass1", "Data type to colour plot:",\n',
    '                           choices = c("Cell Information", paste0("Assay: ", {prefix}def$assay)), \n',
    '                           selected = "Cell Information"), \n',
    '            selectInput("{prefix}s2inp1", "Cell Info / Feature Name:", choices = NULL) %>% \n',
    '              helper(type = "inline", size = "m", fade = TRUE,\n',
    '                     title = "Cell Info / Gene to colour cells by",\n',
    '                     content = c("Select cell info / feature to colour cells",\n',
    '                                 "- Categorical covariates have a fixed colour palette",\n',
    '                                 paste0("- Continuous covariates / gene expression are coloured ",\n',
    '                                        "in a Blue-Yellow-Red colour scheme, which can be ",\n',
    '                                        "changed in the plot controls")))\n',
    '          ),\n',
    '          column(\n',
    '            6, actionButton("{prefix}s2tog1", "Toggle plot controls"),\n',
    '            conditionalPanel(\n',
    '              condition = "input.{prefix}s2tog1 % 2 == 1",\n',
    '              radioButtons("{prefix}s2col1", "Colour (Continuous data):",\n',
    '                           choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"),\n',
    '                           selected = "Blue-Yellow-Red"),\n',
    '              numericInput("{prefix}s2min1", "Min cutoff (q##):",\n',
    '                           min = 0, max = 50, value = 0, step = 1),\n',
    '              numericInput("{prefix}s2max1", "Max cutoff (q##):",\n',
    '                           min = 50, max = 100, value = 100, step = 1),\n',
    '              checkboxInput("{prefix}s2lab1", "Show cell info labels", value = TRUE)\n',
    '            )\n',
    '          )\n',
    '        ),\n',
    '        fluidRow(column(12, uiOutput("{prefix}s2oup1.ui"))),\n',
    '        fluidRow(column(12, uiOutput("{prefix}s2oup3.ui"))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}s2oup1.dl", "Download Plot")),\n',
    '                 column(3, radioButtons("{prefix}s2oup1.f", "Plot format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}s2oup1.h", "Plot height:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)),\n',
    '                 column(3, numericInput("{prefix}s2oup1.w", "Plot width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}s2oup3.dl", "Download Legend")),\n',
    '                 column(3, radioButtons("{prefix}s2oup3.f", "Legend format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}s2oup3.h", "Legend height:",\n',
    '                                        min = 0.2, max = 10, value = 2, step = 0.1)),\n',
    '                 column(3, numericInput("{prefix}s2oup3.w", "Legend width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        br()\n',
    '      ), # End of column (6 space)\n',
    '      column( \n',
    '        6,\n',
    '        fluidRow( \n',
    '          column(\n',
    '            6, selectInput("{prefix}s2ass2", "Data type to colour plot:",\n',
    '                           choices = c("Cell Information", paste0("Assay: ", {prefix}def$assay)), \n',
    '                           selected = paste0("Assay: ", {prefix}def$assay[1])), \n',
    '            selectInput("{prefix}s2inp2", "Cell Info / Feature Name:", choices = NULL) %>% \n',
    '              helper(type = "inline", size = "m", fade = TRUE,\n',
    '                     title = "Cell Info / Gene to colour cells by",\n',
    '                     content = c("Select cell info / feature to colour cells",\n',
    '                                 "- Categorical covariates have a fixed colour palette",\n',
    '                                 paste0("- Continuous covariates / gene expression are coloured ",\n',
    '                                        "in a Blue-Yellow-Red colour scheme, which can be ",\n',
    '                                        "changed in the plot controls")))\n',
    '          ),\n',
    '          column(\n',
    '            6, actionButton("{prefix}s2tog2", "Toggle plot controls"),\n',
    '            conditionalPanel(\n',
    '              condition = "input.{prefix}s2tog2 % 2 == 1",\n',
    '              radioButtons("{prefix}s2col2", "Colour:",\n',
    '                           choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"),\n',
    '                           selected = "White-Red"),\n',
    '              numericInput("{prefix}s2min2", "Min cutoff (q##):",\n',
    '                           min = 0, max = 50, value = 0, step = 1),\n',
    '              numericInput("{prefix}s2max2", "Max cutoff (q##):",\n',
    '                           min = 50, max = 100, value = 100, step = 1),\n',
    '              checkboxInput("{prefix}s2lab2", "Show cell info labels", value = TRUE)\n',
    '            )\n',
    '          )\n',
    '        ) , \n',
    '        fluidRow(column(12, uiOutput("{prefix}s2oup2.ui"))),\n',
    '        fluidRow(column(12, uiOutput("{prefix}s2oup4.ui"))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}s2oup2.dl", "Download Plot")),\n',
    '                 column(3, radioButtons("{prefix}s2oup2.f", "Plot format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}s2oup2.h", "Plot height:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5)),\n',
    '                 column(3, numericInput("{prefix}s2oup2.w", "Plot width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        fluidRow(column(3, downloadButton("{prefix}s2oup4.dl", "Download Legend")),\n',
    '                 column(3, radioButtons("{prefix}s2oup4.f", "Legend format:", \n',
    '                                        choices = c("png", "pdf"), selected = "png", inline = TRUE)),\n',
    '                 column(3, numericInput("{prefix}s2oup4.h", "Legend height:",\n',
    '                                        min = 0.2, max = 10, value = 2, step = 0.1)),\n',
    '                 column(3, numericInput("{prefix}s2oup4.w", "Legend width:",\n',
    '                                        min = 4, max = 20, value = 8, step = 0.5))),\n',
    '        br()\n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  , \n',
    '  \n'
  )
}

#' Write code for ui.R
#'
#' @rdname wrUImainT1
#' @export wrUImainT1
#'
wrUImainT1 <- function(prefix) {
  glue::glue(
    '  ### Tab1.t1: Trackplot \n',
    '  tabPanel( \n',
    '    HTML("Track plot"), \n',
    '    h4("Track plot"), \n',
    '    "In this tab, users can visualise open chromatin regions.", \n',
    '    br(),br(), \n',
    '    fluidRow( \n',
    '      column( \n',
    '        3, style="border-right: 2px solid black", \n',
    '        radioButtons("{prefix}t1sel", "Input method:", choices = c("Gene", "Region"), \n',
    '                     selected = "Region", inline = TRUE),\n',
    '        conditionalPanel(\n',
    '          condition = "input.{prefix}t1sel == \'Gene\'",\n',
    '          selectInput("{prefix}t1gene", "Gene:",\n',
    '                      choices = NULL)),\n',
    '        conditionalPanel(\n',
    '          condition = "input.{prefix}t1sel == \'Region\'",\n',
    '          selectInput("{prefix}t1chr", "Chromosome:",\n',
    '                      choices = {prefix}bw$chrSize$chr, \n',
    '                      selected = {prefix}bw$chrGene[42]$chr), \n',
    '          fluidRow(column(6, textInput("{prefix}t1srt", "Start", value = {prefix}bw$chrGene[42]$srt)),\n',
    '                   column(6, textInput("{prefix}t1end", "End",   value = {prefix}bw$chrGene[42]$end)))\n',
    '        ),\n',
    '        selectInput("{prefix}t1grp", "Group by:",\n',
    '                    choices = {prefix}bwGrp,\n',
    '                    selected = {prefix}bwGrp[1]),\n',
    '        selectInput("{prefix}t1bed", "Track annotation bedfile:", choices = NULL),\n',
    '        fluidRow(column(6, numericInput("{prefix}t1extL", "extend left by:",\n',
    '                                        min = 0, max = 50000, value = 5000, step = 1000)),\n',
    '                 column(6, numericInput("{prefix}t1extR", "extend right by:",\n',
    '                                        min = 0, max = 50000, value = 5000, step = 1000))),\n',
    '        checkboxInput("{prefix}t1col", \n',
    '                      label = "Check to color track plot:", \n',
    '                      value = TRUE), # FALSE means "gray" is selected by default\n',
    '        actionButton("{prefix}t1tog1", "Toggle to subset cells"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}t1tog1 % 2 == 1", \n',
    '          uiOutput("{prefix}t1sub1.ui"),\n',
    '          actionButton("{prefix}t1sub1all", "Select all groups", class = "btn btn-primary"), \n',
    '          actionButton("{prefix}t1sub1non", "Deselect all groups", class = "btn btn-primary") \n',
    '        ), \n',
    '        div(style = "display:inline-block; float:right", actionButton("{prefix}t1update", "Update View")),\n',
    '        br(), br(),\n',
    '        fluidRow(column(6, numericInput("{prefix}t1track.h", "plot height (pixels):",\n',
    '                                        min = 400, max = 2000, value = 600, step = 50)),\n',
    '                 column(6, numericInput("{prefix}t1track.w", "plot width (pixels):",\n',
    '                                        min = 400, max = 2000, value = 1000, step = 50)))\n',
    '      ), \n',
    '      column(\n',
    '        9, actionButton("{prefix}t1tog0", "Toggle graphics controls"), \n',
    '        conditionalPanel( \n',
    '          condition = "input.{prefix}t1tog0 % 2 == 1", \n',
    '          fluidRow( \n',
    '            column(4, \n',
    '                   checkboxInput("{prefix}t1dgt", "Display gene track", value = TRUE),\n',
    '                   checkboxInput("{prefix}t1asg", "Auto scale group", value = TRUE),\n',
    '                   checkboxInput("{prefix}t1sha", "Show axis", value = TRUE),\n',
    '                   checkboxInput("{prefix}t1ctx", "Collapse transcripts", value = TRUE)\n',
    '            ), \n',
    '            column(4,\n',
    '                   sliderInput("{prefix}t1ath", "Annotation track height: ", \n',
    '                               min = 1, max = 10, value = 2, step = 1),\n',
    '                   sliderInput("{prefix}t1gth", "Gene track height: ", \n',
    '                               min = 1, max = 10, value = 4, step = 1)\n',
    '            ),\n',
    '            column(4,\n',
    '                   sliderInput("{prefix}t1tnp", "Track names position: ", \n',
    '                               min = 0, max = 1, value = 0, step = 0.1),\n',
    '                   sliderInput("{prefix}t1gsz", "Font size: ", \n',
    '                               min = 0, max = 3, value = 2, step = 0.25) \n',
    '            ))),\n',
    '        uiOutput("{prefix}t1track.ui")\n',
    '      )  # End of column (6 space) \n',
    '    )    # End of fluidRow (4 space) \n',
    '  )      # End of tab (2 space) \n',
    '  , \n',
    '  \n'
  )
}

#' Write code for ui.R
#' 
#' @param footnote shiny app footnote
#'
#' @rdname wrUIpost
#' @export wrUIpost
#'
wrUIpost <- function(footnote) {
  if(is.list(footnote)){
    f1 = ''; f2 = ''; f3 = ''; f4 = ''; f5 = ''; f6 = ''; f7 = ''; f8 = ''
    if(!is.null(footnote$author)){ f1 = paste0('"',footnote$author,' ",')}
    if(!is.null(footnote$title)){  f2 = paste0('"',footnote$title,' ",')}
    if(!is.null(footnote$journal)){f3 = paste0('em("',footnote$journal,' "),')}
    if(!is.null(footnote$volume)){ f4 = paste0('strong("',footnote$volume,', "),')}
    if(!is.null(footnote$page)){   f5 = paste0('"',footnote$page,' ",')}
    if(!is.null(footnote$year)){   f6 = paste0('"(',footnote$year,') ",')}
    if(!is.null(footnote$doi)){    f7 = paste0('"doi: ',footnote$doi,' ",')}
    if(!is.null(footnote$link)){   f8 = paste0('a("[Link]", href = "',footnote$link,'", target="_blank"),')}
    f0 = paste0('strong("Reference: "),', f1, f2, f3, f4, f5, f6, f7, f8, 
                'style = "font-size: 125%;"')
  } else {
    f0 = paste0('"', footnote[[1]], '", style = "font-size: 125%;"')
  }
  
  glue::glue(
    '    br(), \n',
    '    p({f0}), \n',
    '    p(em("This webpage was made using "), a("ShinyCell2", \n',
    '       href = "https://github.com/the-ouyang-lab/ShinyCell2",target="_blank")), \n',
    '    br(),br(),br(),br(),br()  \n',
    '  )))  \n\n\n'
  )
}

#' Write code for google-analytics.html
#'
#' @param gaID Google analytics tracking ID (e.g. "UA-123456789-0")
#' 
#' @rdname wrUIga
#' @export wrUIga
#'
wrUIga <- function(gaID) {
  glue::glue('<!-- Global site tag (gtag.js) - Google Analytics --> \n',
             '  <script async src="https://www.googletagmanager.com/gtag/js?id={gaID}"></script> \n',
             '  <script> \n',
             '  window.dataLayer = window.dataLayer || []; \n',
             'function gtag(){{dataLayer.push(arguments);}} \n',
             'gtag(\'js\', new Date()); \n',
             ' \n',
             'gtag(\'config\', \'{gaID}\'); \n',
             '</script> \n')
}


